{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cf9f5180-d798-4552-838b-e3ee66d1d863",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ignoring pip: markers 'python_version < \"3\"' don't match your environment\n",
      "Looking in links: /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo/avx2, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo/generic, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/generic\n",
      "Requirement already satisfied: pysam in /localscratch/dtaliun.15256098.0/lib/python3.9/site-packages (0.20.0+computecanada)\n",
      "Ignoring pip: markers 'python_version < \"3\"' don't match your environment\n",
      "Looking in links: /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo/avx2, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo/generic, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/generic\n",
      "Requirement already satisfied: intervaltree in /localscratch/dtaliun.15256098.0/lib/python3.9/site-packages (3.1.0+computecanada)\n",
      "Requirement already satisfied: sortedcontainers<3.0,>=2.0 in /localscratch/dtaliun.15256098.0/lib/python3.9/site-packages (from intervaltree) (2.4.0+computecanada)\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!{sys.executable} -m pip install pysam --no-index\n",
    "!{sys.executable} -m pip install intervaltree --no-index\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4a32ff5c-98f2-478e-bbfc-42b151fc4e8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "from scipy import stats\n",
    "import glob\n",
    "from intervaltree import Interval, IntervalTree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b4369f4-d9ca-493c-9463-4712f76dd1c9",
   "metadata": {},
   "source": [
    "## 1. Help functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5ab725dc-4040-47b6-aafd-9ad915371eee",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "HAPPY_COLUMNS = ['Type', 'Filter', 'TRUTH.TP', 'QUERY.FP', 'TRUTH.FN', 'METRIC.Recall', 'METRIC.Precision', 'METRIC.F1_Score']\n",
    "\n",
    "def load_targets_bed(filename):\n",
    "    targets = dict()\n",
    "    with open(filename, 'rt') as ifile:\n",
    "        for line in ifile:\n",
    "            chrom, start, stop = line.rstrip().split()\n",
    "            targets.setdefault(chrom, IntervalTree()).addi(int(start) + 1, int(stop) + 1) # because BED is 0-based and IntervalTree stores half-open intervals.\n",
    "    return targets\n",
    "\n",
    "def read_af(filename):\n",
    "    with pysam.VariantFile(filename, 'r') as ifile:\n",
    "        for record in ifile:\n",
    "            if 'ASJ_AF' not in record.info or 'ASJ_AN' not in record.info:\n",
    "                continue\n",
    "            if 'TOMPED_AF' not in record.info:\n",
    "                continue\n",
    "            if len(record.alts) > 1: # skip multi-allelic\n",
    "                continue\n",
    "            if record.info['ASJ_AN'] < 1000: # keep only those with sufficient ASJ sample size in gnomad \n",
    "                continue\n",
    "            asj_af = record.info.get('ASJ_AF', [0.0])[0]\n",
    "            if (asj_af <= 0) or (asj_af >= 1.0): # remove those which are monomorphic in ASJ according to gnomad\n",
    "                continue\n",
    "            topmed_af = record.info.get('TOMPED_AF', [0.0])[0]\n",
    "            if (topmed_af <= 0) or (topmed_af >= 1.0): # remove those which are monomorphic in TOPMed\n",
    "                continue\n",
    "            if record.samples['QUERY']['BD'] == 'UNK' or record.samples['TRUTH']['BD'] == 'UNK': # variants outside the truth set regions\n",
    "                continue\n",
    "            if record.samples['TRUTH']['BD'] == 'N': # skip halfcalls such as \"./1\"\n",
    "                continue\n",
    "            targets = WEGS_TARGETS[record.chrom]\n",
    "            entry = {\n",
    "                     'Name': f'{record.chrom}:{record.pos}:{record.ref}:{record.alts[0]}',\n",
    "                     'ASJ_AF': asj_af,\n",
    "                     'TOPMED_AF': topmed_af,\n",
    "                     'FOLD_CHANGE_AF': asj_af / topmed_af,\n",
    "                     'IN_TARGET': targets.overlaps(record.start - 1) or targets.overlaps(record.stop - 1),\n",
    "                     'AC_LESS': False,\n",
    "                     'AC_GREATER': False,\n",
    "                     'AC_SAME': False\n",
    "                    }\n",
    "            if record.samples['TRUTH']['BD'] == '.':\n",
    "                truth_gt = 0\n",
    "            else:\n",
    "                truth_gt = sum(record.samples['TRUTH']['GT'])\n",
    "            if record.samples['QUERY']['BD'] == '.':\n",
    "                query_gt = 0\n",
    "            else:\n",
    "                query_gt = sum(record.samples['QUERY']['GT'])            \n",
    "            if truth_gt > query_gt:\n",
    "                entry['AC_LESS'] = True\n",
    "            elif truth_gt < query_gt:\n",
    "                entry['AC_GREATER'] = True\n",
    "            else:\n",
    "                entry['AC_SAME'] = True\n",
    "            yield(entry)\n",
    "            \n",
    "def read_bad_variant_calls(filename):\n",
    "    with pysam.VariantFile(filename, 'r') as ifile:\n",
    "        for record in ifile:\n",
    "            if 'ASJ_AF' not in record.info or 'ASJ_AN' not in record.info:\n",
    "                continue\n",
    "            if 'TOMPED_AF' not in record.info:\n",
    "                continue\n",
    "            if len(record.alts) > 1: # skip multi-allelic\n",
    "                continue\n",
    "            if record.info['ASJ_AN'] < 1000: # keep only those with sufficient ASJ sample size in gnomad \n",
    "                continue\n",
    "            asj_af = record.info.get('ASJ_AF', [0.0])[0]\n",
    "            if (asj_af <= 0) or (asj_af >= 1.0): # remove those which are monomorphic in ASJ according to gnomad\n",
    "                continue\n",
    "            topmed_af = record.info.get('TOMPED_AF', [0.0])[0]\n",
    "            if (topmed_af <= 0) or (topmed_af >= 1.0): # remove those which are monomorphic in TOPMed\n",
    "                continue\n",
    "            if record.samples['QUERY']['BD'] == 'UNK' or record.samples['TRUTH']['BD'] == 'UNK': # variants outside the truth set regions\n",
    "                continue\n",
    "            if record.samples['TRUTH']['BD'] == 'N': # skip halfcalls such as \"./1\"\n",
    "                continue\n",
    "            if record.samples['TRUTH']['BD'] == '.':\n",
    "                truth_gt = 0\n",
    "            else:\n",
    "                truth_gt = sum(record.samples['TRUTH']['GT'])\n",
    "            if record.samples['QUERY']['BD'] == '.':\n",
    "                query_gt = 0\n",
    "            else:\n",
    "                query_gt = sum(record.samples['QUERY']['GT'])            \n",
    "            if truth_gt != query_gt:\n",
    "                yield f'{record.chrom}:{record.pos}:{record.ref}:{record.alts[0]}'\n",
    "                \n",
    "def read_tp_snvs(filename):\n",
    "    with pysam.VariantFile(filename, 'r') as ifile:\n",
    "        for record in ifile.fetch():\n",
    "            if len(record.alts) > 1: # skip multi-allelic\n",
    "                continue\n",
    "            if len(record.ref) != 1 or len(record.alts[0]) != 1: # keep only SNVs\n",
    "                continue\n",
    "            if record.samples['QUERY']['BD'] == 'UNK' or record.samples['TRUTH']['BD'] == 'UNK': # variants outside the truth set regions\n",
    "                continue\n",
    "            if record.samples['TRUTH']['BD'] == 'N': # skip halfcalls such as \"./1\"\n",
    "                continue\n",
    "            targets = WEGS_TARGETS[record.chrom]\n",
    "            if not (targets.overlaps(record.start - 1) or targets.overlaps(record.stop - 1)): # use only variants within target regions\n",
    "                continue\n",
    "            if record.samples['TRUTH']['BD'] == '.':\n",
    "                truth_gt = 0\n",
    "            else:\n",
    "                truth_gt = sum(record.samples['TRUTH']['GT'])\n",
    "            if record.samples['QUERY']['BD'] == '.':\n",
    "                query_gt = 0\n",
    "            else:\n",
    "                query_gt = sum(record.samples['QUERY']['GT'])            \n",
    "            if truth_gt != query_gt: # skip not TP\n",
    "                continue\n",
    "            yield f'{record.chrom}:{record.pos}:{record.ref}:{record.alts[0]}'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec4bed38-36a4-437d-9608-65c30e55a358",
   "metadata": {},
   "source": [
    "## 2. Input files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "051f9bff-7cba-49bb-9d57-00edc75d373d",
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE_ANNOTATIONS_FILE = 'sample_annotations.txt'\n",
    "WEGS_TARGETS_BED_FILE = 'WES_targets/SureSelectHumanAllExonV7.b37.Target.bed'\n",
    "TOPMED_ARRAY_BENCHMARK_DIR = 'TOPMed_array_imputation_analysis'\n",
    "WEGS_BENCHMARK_DIR = 'Variant_calling_analysis/WEGS/'\n",
    "TP_ONLY_IN_WEGS_DIR = 'TOPMed_array_imputation_analysis/Only_in_WEGS'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6dac4326-c1da-4ad0-a474-c08d19fdc8e8",
   "metadata": {},
   "source": [
    "## 3. Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e782394a-a3ce-42b5-87fc-a36a5ec3dba5",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load sample annotations\n",
    "df_samples = pd.read_csv(SAMPLE_ANNOTATIONS_FILE, header = 0, sep = '\\t')\n",
    "\n",
    "hgid2naid = {\n",
    "    'HG002': 'NA24385',\n",
    "    'HG003': 'NA24149',\n",
    "    'HG004': 'NA24143'\n",
    "}\n",
    "\n",
    "hgid2rel = {\n",
    "    'HG002': 'Son',\n",
    "    'HG003': 'Father',\n",
    "    'HG004': 'Mother'\n",
    "}\n",
    "\n",
    "WEGS_TARGETS = load_targets_bed(WEGS_TARGETS_BED_FILE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6794ef2-7250-4918-9e34-206699a57ff5",
   "metadata": {},
   "source": [
    "### 3.1 Load TOPMed imputed variants from array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9e4a671f-8fe0-47e3-92a9-7fd50eee97f4",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_array_imputed_variants = []\n",
    "for row_index, row in df_samples[['HGID', 'Family']].drop_duplicates().iterrows():\n",
    "    filepath = f'{TOPMED_ARRAY_BENCHMARK_DIR}/Genomewide/{row.HGID}_{hgid2naid[row.HGID]}_{row.Family}.dose.b37.happy_benchmark.summary.csv'\n",
    "    df = pd.read_csv(filepath, usecols = HAPPY_COLUMNS).rename(columns = {\n",
    "        'METRIC.Recall': 'Recall', \n",
    "        'METRIC.Precision': 'Precision',\n",
    "        'METRIC.F1_Score': 'F1_Score',\n",
    "        'TRUTH.TP': 'TP',\n",
    "        'QUERY.FP': 'FP',\n",
    "        'TRUTH.FN': 'FN'\n",
    "    })\n",
    "    df['Total_called'] = df['TP'] + df['FP']\n",
    "    df['HGID'] = row.HGID\n",
    "    df_array_imputed_variants.append(df)\n",
    "df_array_imputed_variants = pd.concat(df_array_imputed_variants).reset_index(drop = True)\n",
    "\n",
    "df_array_imputed_target_variants = []\n",
    "for row_index, row in df_samples[['HGID', 'Family']].drop_duplicates().iterrows():\n",
    "    filepath = f'{TOPMED_ARRAY_BENCHMARK_DIR}/Targets/{row.HGID}_{hgid2naid[row.HGID]}_{row.Family}.dose.b37.happy_benchmark.summary.csv'\n",
    "    df = pd.read_csv(filepath, usecols = HAPPY_COLUMNS).rename(columns = {\n",
    "        'METRIC.Recall': 'Recall', \n",
    "        'METRIC.Precision': 'Precision',\n",
    "        'METRIC.F1_Score': 'F1_Score',\n",
    "        'TRUTH.TP': 'TP',\n",
    "        'QUERY.FP': 'FP',\n",
    "        'TRUTH.FN': 'FN'\n",
    "    })\n",
    "    df['Total_called'] = df['TP'] + df['FP']\n",
    "    df['HGID'] = row.HGID\n",
    "    df_array_imputed_target_variants.append(df)\n",
    "df_array_imputed_target_variants = pd.concat(df_array_imputed_target_variants).reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "566e6942-2652-4275-afde-3b70ac439ab7",
   "metadata": {},
   "source": [
    "### 3.2 Load WEGS variant calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "20537a04-83ea-44cc-b4d5-ff6c5a044935",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_wegs_variants = []\n",
    "for row_index, row in df_samples.iterrows():\n",
    "    if row.Plexing == 1:\n",
    "        continue\n",
    "    for wgs_depth in [2, 5]:\n",
    "        filepath = f'{WEGS_BENCHMARK_DIR}/{row.Plexing}plex_{wgs_depth}X_genomewide/{row.Filename}.norm.hard_filter.happy_benchmark.summary.csv'\n",
    "        df = pd.read_csv(filepath, usecols = HAPPY_COLUMNS).rename(columns = {\n",
    "            'METRIC.Recall': 'Recall', \n",
    "            'METRIC.Precision': 'Precision',\n",
    "            'METRIC.F1_Score': 'F1_Score',\n",
    "            'TRUTH.TP': 'TP',\n",
    "            'QUERY.FP': 'FP',\n",
    "            'TRUTH.FN': 'FN'\n",
    "        })\n",
    "        df['Total_called'] = df['TP'] + df['FP']\n",
    "        df['ID'] = row.Filename\n",
    "        df['HGID'] = row.HGID\n",
    "        df['Batch'] = row.Batch\n",
    "        df['Plexing'] = row.Plexing\n",
    "        df['WGS_depth'] = wgs_depth\n",
    "        df_wegs_variants.append(df)\n",
    "df_wegs_variants = pd.concat(df_wegs_variants).reset_index(drop = True)\n",
    "\n",
    "df_wegs_target_variants = []\n",
    "for row_index, row in df_samples.iterrows():\n",
    "    if row.Plexing == 1:\n",
    "        continue\n",
    "    for wgs_depth in [2, 5]:\n",
    "        filepath = f'{WEGS_BENCHMARK_DIR}/{row.Plexing}plex_{wgs_depth}X/{row.Filename}.norm.hard_filter.happy_benchmark.summary.csv'\n",
    "        df = pd.read_csv(filepath, usecols = HAPPY_COLUMNS).rename(columns = {\n",
    "            'METRIC.Recall': 'Recall', \n",
    "            'METRIC.Precision': 'Precision',\n",
    "            'METRIC.F1_Score': 'F1_Score',\n",
    "            'TRUTH.TP': 'TP',\n",
    "            'QUERY.FP': 'FP',\n",
    "            'TRUTH.FN': 'FN'\n",
    "        })\n",
    "        df['Total_called'] = df['TP'] + df['FP']\n",
    "        df['ID'] = row.Filename\n",
    "        df['HGID'] = row.HGID\n",
    "        df['Batch'] = row.Batch\n",
    "        df['Plexing'] = row.Plexing\n",
    "        df['WGS_depth'] = wgs_depth\n",
    "        df_wegs_target_variants.append(df)\n",
    "df_wegs_target_variants = pd.concat(df_wegs_target_variants).reset_index(drop = True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9f963dc-7f9a-45be-bfcf-3f8e985a1c8a",
   "metadata": {},
   "source": [
    "### 3.3 Load variant calls annotated with allele frequencies from TOPMed and gnomAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "121b9a9b-3edc-4ea3-8368-cf18e941e7ce",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'TOPMed_array_imputation_analysis/Genomewide_with_AF/HG002_NA24385_Son.dose.b37.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'TOPMed_array_imputation_analysis/Genomewide_with_AF/HG003_NA24149_Father.dose.b37.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'TOPMed_array_imputation_analysis/Genomewide_with_AF/HG004_NA24143_Mother.dose.b37.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//4plex_2X_genomewide_with_AF/HG002_NA24385_Son_SSXTHSV27054-SSXTHSV25054.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//4plex_2X_genomewide_with_AF/HG002_NA24385_Son_NewIndex_SSXTHSV27061-SSXTHSV25061.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//4plex_2X_genomewide_with_AF/HG002_NA24385_Son_SSXTHSV27050-SSXTHSV25050.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//4plex_2X_genomewide_with_AF/HG002_NA24385_Son_SSXTHSV27058-SSXTHSV25058.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//4plex_2X_genomewide_with_AF/HG003_NA24149_Father_NewIndex_SSXTHSV27053-SSXTHSV25053.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//4plex_2X_genomewide_with_AF/HG003_NA24149_Father_SSXTHSV27056-SSXTHSV25056.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//4plex_2X_genomewide_with_AF/HG003_NA24149_Father_SSXTHSV27060-SSXTHSV25060.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//4plex_2X_genomewide_with_AF/HG003_NA24149_Father_SSXTHSV27052-SSXTHSV25052.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//4plex_2X_genomewide_with_AF/HG004_NA24143_Mother_SSXTHSV27059-SSXTHSV25059.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//4plex_2X_genomewide_with_AF/HG004_NA24143_Mother_SSXTHSV27051-SSXTHSV25051.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//4plex_2X_genomewide_with_AF/HG004_NA24143_Mother_SSXTHSV27055-SSXTHSV25055.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//4plex_2X_genomewide_with_AF/HG004_NA24143_Mother_NewIndex_SSXTHSV27057-SSXTHSV25057.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG002_NA24385_Son_SSXTHSV27062-SSXTHSV25062.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG002_NA24385_Son_NewIndex_SSXTHSV27077-SSXTHSV25077.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG002_NA24385_Son_NewIndex_SSXTHSV27063-SSXTHSV25063.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG002_NA24385_Son_SSXTHSV27081-SSXTHSV25081.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG002_NA24385_Son_NewIndex_SSXTHSV27071-SSXTHSV25071.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG003_NA24149_Father_NewIndex_SSXTHSV27075-SSXTHSV25075.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG003_NA24149_Father_NewIndex_SSXTHSV27068-SSXTHSV25068.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG003_NA24149_Father_NewIndex_SSXTHSV27076-SSXTHSV25076.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG003_NA24149_Father_SSXTHSV27074-SSXTHSV25074.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG003_NA24149_Father_NewIndex_SSXTHSV27069-SSXTHSV25069.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG003_NA24149_Father_SSXTHSV27067-SSXTHSV25067.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG004_NA24143_Mother_SSXTHSV27064-SSXTHSV25064.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG004_NA24143_Mother_SSXTHSV27072-SSXTHSV25072.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG004_NA24143_Mother_NewIndex_SSXTHSV27073-SSXTHSV25073.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG004_NA24143_Mother_NewIndex_SSXTHSV27066-SSXTHSV25066.norm.hard_filter.happy_benchmark.vcf.gz'\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'Variant_calling_analysis/WEGS//8plex_5X_genomewide_with_AF/HG004_NA24143_Mother_NewIndex_SSXTHSV27065-SSXTHSV25065.norm.hard_filter.happy_benchmark.vcf.gz'\n"
     ]
    }
   ],
   "source": [
    "imputed_arrays = {}\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    filename = f'{TOPMED_ARRAY_BENCHMARK_DIR}/Genomewide_with_AF/{hgid}_{naid}_{hgid2rel[hgid]}.dose.b37.happy_benchmark.vcf.gz'\n",
    "    imputed_arrays[hgid] = pd.DataFrame(read_af(filename))\n",
    "    \n",
    "wegs_bad_4p2x = dict()\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    wegs_bad_4p2x[hgid] = []\n",
    "    for filename in glob.glob(f'{WEGS_BENCHMARK_DIR}/4plex_2X_genomewide_with_AF/{hgid}_{naid}_*.norm.hard_filter.happy_benchmark.vcf.gz'):\n",
    "        wegs_bad_4p2x[hgid].append(set(read_bad_variant_calls(filename)))\n",
    "\n",
    "wegs_bad_8p5x = dict()\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    wegs_bad_8p5x[hgid] = []\n",
    "    for filename in glob.glob(f'{WEGS_BENCHMARK_DIR}/8plex_5X_genomewide_with_AF/{hgid}_{naid}_*.norm.hard_filter.happy_benchmark.vcf.gz'):\n",
    "        wegs_bad_8p5x[hgid].append(set(read_bad_variant_calls(filename)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b09c7ae4-790b-421c-8041-3d956edddb4b",
   "metadata": {},
   "source": [
    "## 4. Figures and tables"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a0e965f-4f2c-4347-91e2-284e88032f77",
   "metadata": {},
   "source": [
    "### TABLE: Precision and recall rates of variants imputed using the TOPMed reference panel inside WES target regions (PART 1 - SNVs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4d5c3e36-0278-4c96-9ddb-c6df6ae80d34",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Imputed TP</th>\n",
       "      <th>Imputed FN</th>\n",
       "      <th>Imputed FP</th>\n",
       "      <th>Imputed Precision</th>\n",
       "      <th>Imputed Recall</th>\n",
       "      <th>WEGS 4P2X TP</th>\n",
       "      <th>WEGS 4P2X FN</th>\n",
       "      <th>WEGS 4P2X FP</th>\n",
       "      <th>WEGS 4P2X Precision</th>\n",
       "      <th>WEGS 4P2X Recall</th>\n",
       "      <th>WEGS 8P5X TP</th>\n",
       "      <th>WEGS 8P5X FN</th>\n",
       "      <th>WEGS 8P5X FP</th>\n",
       "      <th>WEGS 8P5X Precision</th>\n",
       "      <th>WEGS 8P5X Recall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002</td>\n",
       "      <td>21,458</td>\n",
       "      <td>1,285</td>\n",
       "      <td>165</td>\n",
       "      <td>0.9924</td>\n",
       "      <td>0.9435</td>\n",
       "      <td>22,379 (7)</td>\n",
       "      <td>364 (7)</td>\n",
       "      <td>446 (6)</td>\n",
       "      <td>0.9805 (0.0002)</td>\n",
       "      <td>0.9840 (0.0003)</td>\n",
       "      <td>22,390 (5)</td>\n",
       "      <td>353 (5)</td>\n",
       "      <td>449 (7)</td>\n",
       "      <td>0.9803 (0.0003)</td>\n",
       "      <td>0.9845 (0.0002)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG003</td>\n",
       "      <td>21,477</td>\n",
       "      <td>1,112</td>\n",
       "      <td>111</td>\n",
       "      <td>0.9949</td>\n",
       "      <td>0.9508</td>\n",
       "      <td>22,231 (9)</td>\n",
       "      <td>358 (9)</td>\n",
       "      <td>425 (4)</td>\n",
       "      <td>0.9813 (0.0001)</td>\n",
       "      <td>0.9841 (0.0004)</td>\n",
       "      <td>22,249 (6)</td>\n",
       "      <td>340 (6)</td>\n",
       "      <td>406 (5)</td>\n",
       "      <td>0.9821 (0.0002)</td>\n",
       "      <td>0.9850 (0.0003)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG004</td>\n",
       "      <td>21,228</td>\n",
       "      <td>1,315</td>\n",
       "      <td>181</td>\n",
       "      <td>0.9915</td>\n",
       "      <td>0.9417</td>\n",
       "      <td>22,195 (9)</td>\n",
       "      <td>348 (9)</td>\n",
       "      <td>406 (7)</td>\n",
       "      <td>0.9820 (0.0003)</td>\n",
       "      <td>0.9846 (0.0004)</td>\n",
       "      <td>22,197 (5)</td>\n",
       "      <td>346 (5)</td>\n",
       "      <td>402 (3)</td>\n",
       "      <td>0.9822 (0.0001)</td>\n",
       "      <td>0.9847 (0.0002)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Imputed TP Imputed FN Imputed FP Imputed Precision Imputed Recall  \\\n",
       "0  HG002     21,458      1,285        165            0.9924         0.9435   \n",
       "1  HG003     21,477      1,112        111            0.9949         0.9508   \n",
       "2  HG004     21,228      1,315        181            0.9915         0.9417   \n",
       "\n",
       "  WEGS 4P2X TP WEGS 4P2X FN WEGS 4P2X FP WEGS 4P2X Precision WEGS 4P2X Recall  \\\n",
       "0   22,379 (7)      364 (7)      446 (6)     0.9805 (0.0002)  0.9840 (0.0003)   \n",
       "1   22,231 (9)      358 (9)      425 (4)     0.9813 (0.0001)  0.9841 (0.0004)   \n",
       "2   22,195 (9)      348 (9)      406 (7)     0.9820 (0.0003)  0.9846 (0.0004)   \n",
       "\n",
       "  WEGS 8P5X TP WEGS 8P5X FN WEGS 8P5X FP WEGS 8P5X Precision WEGS 8P5X Recall  \n",
       "0   22,390 (5)      353 (5)      449 (7)     0.9803 (0.0003)  0.9845 (0.0002)  \n",
       "1   22,249 (6)      340 (6)      406 (5)     0.9821 (0.0002)  0.9850 (0.0003)  \n",
       "2   22,197 (5)      346 (5)      402 (3)     0.9822 (0.0001)  0.9847 (0.0002)  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "    'Sample': [],\n",
    "    'Imputed TP': [],\n",
    "    'Imputed FN': [],\n",
    "    'Imputed FP': [],\n",
    "    'Imputed Precision': [],\n",
    "    'Imputed Recall': [],\n",
    "    'WEGS 4P2X TP': [],\n",
    "    'WEGS 4P2X FN': [],\n",
    "    'WEGS 4P2X FP': [],\n",
    "    'WEGS 4P2X Precision': [],\n",
    "    'WEGS 4P2X Recall': [],\n",
    "    'WEGS 8P5X TP': [],\n",
    "    'WEGS 8P5X FN': [],\n",
    "    'WEGS 8P5X FP': [],\n",
    "    'WEGS 8P5X Precision': [],\n",
    "    'WEGS 8P5X Recall': [],\n",
    "}\n",
    "\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    data['Sample'].append(hgid)\n",
    "    \n",
    "    df_temp = df_array_imputed_target_variants[df_array_imputed_target_variants.HGID == hgid]\n",
    "    \n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL')][['TP', 'FN', 'FP', 'Recall', 'Precision']]\n",
    "    \n",
    "    data['Imputed TP'].append(f'{df_temp.TP.values[0]:,}')\n",
    "    data['Imputed FP'].append(f'{df_temp.FP.values[0]:,}')\n",
    "    data['Imputed FN'].append(f'{df_temp.FN.values[0]:,}')\n",
    "    data['Imputed Precision'].append(f'{df_temp.Precision.values[0]:.4f}')\n",
    "    data['Imputed Recall'].append(f'{df_temp.Recall.values[0]:.4f}')\n",
    "\n",
    "    df_temp = df_wegs_target_variants[df_wegs_target_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 4) & (df_temp.WGS_depth == 2)]\n",
    "    data['WEGS 4P2X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 4P2X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 4P2X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 4P2X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 4P2X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    \n",
    "    df_temp = df_wegs_target_variants[df_wegs_target_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 8) & (df_temp.WGS_depth == 5)]\n",
    "    data['WEGS 8P5X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 8P5X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 8P5X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 8P5X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 8P5X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    \n",
    "pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65c5c901-a49b-4781-916d-0680487e002f",
   "metadata": {},
   "source": [
    "### TABLE: Precision and recall rates of variants imputed using the TOPMed reference panel inside WES target regions (PART 2 - InDels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "240e7740-7ad7-4712-9d3b-24f66036e7c5",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Imputed TP</th>\n",
       "      <th>Imputed FN</th>\n",
       "      <th>Imputed FP</th>\n",
       "      <th>Imputed Precision</th>\n",
       "      <th>Imputed Recall</th>\n",
       "      <th>WEGS 4P2X TP</th>\n",
       "      <th>WEGS 4P2X FN</th>\n",
       "      <th>WEGS 4P2X FP</th>\n",
       "      <th>WEGS 4P2X Precision</th>\n",
       "      <th>WEGS 4P2X Recall</th>\n",
       "      <th>WEGS 8P5X TP</th>\n",
       "      <th>WEGS 8P5X FN</th>\n",
       "      <th>WEGS 8P5X FP</th>\n",
       "      <th>WEGS 8P5X Precision</th>\n",
       "      <th>WEGS 8P5X Recall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002</td>\n",
       "      <td>271</td>\n",
       "      <td>411</td>\n",
       "      <td>8</td>\n",
       "      <td>0.9713</td>\n",
       "      <td>0.3974</td>\n",
       "      <td>648 (3)</td>\n",
       "      <td>34 (3)</td>\n",
       "      <td>57 (1)</td>\n",
       "      <td>0.9197 (0.0017)</td>\n",
       "      <td>0.9498 (0.0050)</td>\n",
       "      <td>647 (2)</td>\n",
       "      <td>35 (2)</td>\n",
       "      <td>58 (3)</td>\n",
       "      <td>0.9177 (0.0038)</td>\n",
       "      <td>0.9481 (0.0028)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG003</td>\n",
       "      <td>257</td>\n",
       "      <td>433</td>\n",
       "      <td>8</td>\n",
       "      <td>0.9698</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>649 (2)</td>\n",
       "      <td>41 (2)</td>\n",
       "      <td>52 (2)</td>\n",
       "      <td>0.9254 (0.0019)</td>\n",
       "      <td>0.9409 (0.0027)</td>\n",
       "      <td>651 (1)</td>\n",
       "      <td>39 (1)</td>\n",
       "      <td>52 (1)</td>\n",
       "      <td>0.9265 (0.0018)</td>\n",
       "      <td>0.9432 (0.0019)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG004</td>\n",
       "      <td>259</td>\n",
       "      <td>384</td>\n",
       "      <td>7</td>\n",
       "      <td>0.9738</td>\n",
       "      <td>0.4028</td>\n",
       "      <td>616 (2)</td>\n",
       "      <td>28 (2)</td>\n",
       "      <td>43 (2)</td>\n",
       "      <td>0.9357 (0.0035)</td>\n",
       "      <td>0.9572 (0.0026)</td>\n",
       "      <td>615 (2)</td>\n",
       "      <td>28 (2)</td>\n",
       "      <td>40 (2)</td>\n",
       "      <td>0.9393 (0.0033)</td>\n",
       "      <td>0.9568 (0.0031)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Imputed TP Imputed FN Imputed FP Imputed Precision Imputed Recall  \\\n",
       "0  HG002        271        411          8            0.9713         0.3974   \n",
       "1  HG003        257        433          8            0.9698         0.3725   \n",
       "2  HG004        259        384          7            0.9738         0.4028   \n",
       "\n",
       "  WEGS 4P2X TP WEGS 4P2X FN WEGS 4P2X FP WEGS 4P2X Precision WEGS 4P2X Recall  \\\n",
       "0      648 (3)       34 (3)       57 (1)     0.9197 (0.0017)  0.9498 (0.0050)   \n",
       "1      649 (2)       41 (2)       52 (2)     0.9254 (0.0019)  0.9409 (0.0027)   \n",
       "2      616 (2)       28 (2)       43 (2)     0.9357 (0.0035)  0.9572 (0.0026)   \n",
       "\n",
       "  WEGS 8P5X TP WEGS 8P5X FN WEGS 8P5X FP WEGS 8P5X Precision WEGS 8P5X Recall  \n",
       "0      647 (2)       35 (2)       58 (3)     0.9177 (0.0038)  0.9481 (0.0028)  \n",
       "1      651 (1)       39 (1)       52 (1)     0.9265 (0.0018)  0.9432 (0.0019)  \n",
       "2      615 (2)       28 (2)       40 (2)     0.9393 (0.0033)  0.9568 (0.0031)  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "    'Sample': [],\n",
    "    'Imputed TP': [],\n",
    "    'Imputed FN': [],\n",
    "    'Imputed FP': [],\n",
    "    'Imputed Precision': [],\n",
    "    'Imputed Recall': [],\n",
    "    'WEGS 4P2X TP': [],\n",
    "    'WEGS 4P2X FN': [],\n",
    "    'WEGS 4P2X FP': [],\n",
    "    'WEGS 4P2X Precision': [],\n",
    "    'WEGS 4P2X Recall': [],\n",
    "    'WEGS 8P5X TP': [],\n",
    "    'WEGS 8P5X FN': [],\n",
    "    'WEGS 8P5X FP': [],\n",
    "    'WEGS 8P5X Precision': [],\n",
    "    'WEGS 8P5X Recall': [],\n",
    "}\n",
    "\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    data['Sample'].append(hgid)\n",
    "    \n",
    "    df_temp = df_array_imputed_target_variants[df_array_imputed_target_variants.HGID == hgid]\n",
    "    \n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL')][['TP', 'FN', 'FP', 'Recall', 'Precision']]\n",
    "    \n",
    "    data['Imputed TP'].append(f'{df_temp.TP.values[0]:,}')\n",
    "    data['Imputed FP'].append(f'{df_temp.FP.values[0]:,}')\n",
    "    data['Imputed FN'].append(f'{df_temp.FN.values[0]:,}')\n",
    "    data['Imputed Precision'].append(f'{df_temp.Precision.values[0]:.4f}')\n",
    "    data['Imputed Recall'].append(f'{df_temp.Recall.values[0]:.4f}')\n",
    "\n",
    "    df_temp = df_wegs_target_variants[df_wegs_target_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 4) & (df_temp.WGS_depth == 2)]\n",
    "    data['WEGS 4P2X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 4P2X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 4P2X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 4P2X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 4P2X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    \n",
    "    df_temp = df_wegs_target_variants[df_wegs_target_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 8) & (df_temp.WGS_depth == 5)]\n",
    "    data['WEGS 8P5X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 8P5X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 8P5X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 8P5X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 8P5X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    \n",
    "pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4529249b-5cb6-4e06-bc6f-e4676e06f0b0",
   "metadata": {},
   "source": [
    "### TABLE: Precision and recall rates of variants imputed using the TOPMed reference panel genome-wide (PART 1 - SNVs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0644311a-2f6a-4cdd-8525-991b325b55ae",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Imputed TP</th>\n",
       "      <th>Imputed FN</th>\n",
       "      <th>Imputed FP</th>\n",
       "      <th>Imputed Precision</th>\n",
       "      <th>Imputed Recall</th>\n",
       "      <th>WEGS 4P2X TP</th>\n",
       "      <th>WEGS 4P2X FN</th>\n",
       "      <th>WEGS 4P2X FP</th>\n",
       "      <th>WEGS 4P2X Precision</th>\n",
       "      <th>WEGS 4P2X Recall</th>\n",
       "      <th>WEGS 8P5X TP</th>\n",
       "      <th>WEGS 8P5X FN</th>\n",
       "      <th>WEGS 8P5X FP</th>\n",
       "      <th>WEGS 8P5X Precision</th>\n",
       "      <th>WEGS 8P5X Recall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002</td>\n",
       "      <td>3,212,631</td>\n",
       "      <td>140,048</td>\n",
       "      <td>21,239</td>\n",
       "      <td>0.9934</td>\n",
       "      <td>0.9582</td>\n",
       "      <td>1,323,307 (13,949)</td>\n",
       "      <td>2,029,372 (13,949)</td>\n",
       "      <td>293,787 (1,558)</td>\n",
       "      <td>0.8183 (0.0023)</td>\n",
       "      <td>0.3947 (0.0042)</td>\n",
       "      <td>1,873,500 (1,770)</td>\n",
       "      <td>1,479,179 (1,770)</td>\n",
       "      <td>242,833 (353)</td>\n",
       "      <td>0.8853 (0.0002)</td>\n",
       "      <td>0.5588 (0.0005)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG003</td>\n",
       "      <td>3,181,715</td>\n",
       "      <td>132,428</td>\n",
       "      <td>17,060</td>\n",
       "      <td>0.9947</td>\n",
       "      <td>0.9600</td>\n",
       "      <td>1,518,733 (9,194)</td>\n",
       "      <td>1,795,410 (9,194)</td>\n",
       "      <td>325,169 (886)</td>\n",
       "      <td>0.8236 (0.0013)</td>\n",
       "      <td>0.4583 (0.0028)</td>\n",
       "      <td>2,151,757 (2,243)</td>\n",
       "      <td>1,162,386 (2,243)</td>\n",
       "      <td>228,413 (270)</td>\n",
       "      <td>0.9040 (0.0002)</td>\n",
       "      <td>0.6493 (0.0007)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG004</td>\n",
       "      <td>3,202,593</td>\n",
       "      <td>137,878</td>\n",
       "      <td>20,523</td>\n",
       "      <td>0.9936</td>\n",
       "      <td>0.9587</td>\n",
       "      <td>1,159,478 (6,227)</td>\n",
       "      <td>2,180,992 (6,227)</td>\n",
       "      <td>291,399 (438)</td>\n",
       "      <td>0.7991 (0.0011)</td>\n",
       "      <td>0.3471 (0.0019)</td>\n",
       "      <td>1,671,569 (3,941)</td>\n",
       "      <td>1,668,902 (3,941)</td>\n",
       "      <td>253,243 (540)</td>\n",
       "      <td>0.8684 (0.0005)</td>\n",
       "      <td>0.5004 (0.0012)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Imputed TP Imputed FN Imputed FP Imputed Precision Imputed Recall  \\\n",
       "0  HG002  3,212,631    140,048     21,239            0.9934         0.9582   \n",
       "1  HG003  3,181,715    132,428     17,060            0.9947         0.9600   \n",
       "2  HG004  3,202,593    137,878     20,523            0.9936         0.9587   \n",
       "\n",
       "         WEGS 4P2X TP        WEGS 4P2X FN     WEGS 4P2X FP  \\\n",
       "0  1,323,307 (13,949)  2,029,372 (13,949)  293,787 (1,558)   \n",
       "1   1,518,733 (9,194)   1,795,410 (9,194)    325,169 (886)   \n",
       "2   1,159,478 (6,227)   2,180,992 (6,227)    291,399 (438)   \n",
       "\n",
       "  WEGS 4P2X Precision WEGS 4P2X Recall       WEGS 8P5X TP       WEGS 8P5X FN  \\\n",
       "0     0.8183 (0.0023)  0.3947 (0.0042)  1,873,500 (1,770)  1,479,179 (1,770)   \n",
       "1     0.8236 (0.0013)  0.4583 (0.0028)  2,151,757 (2,243)  1,162,386 (2,243)   \n",
       "2     0.7991 (0.0011)  0.3471 (0.0019)  1,671,569 (3,941)  1,668,902 (3,941)   \n",
       "\n",
       "    WEGS 8P5X FP WEGS 8P5X Precision WEGS 8P5X Recall  \n",
       "0  242,833 (353)     0.8853 (0.0002)  0.5588 (0.0005)  \n",
       "1  228,413 (270)     0.9040 (0.0002)  0.6493 (0.0007)  \n",
       "2  253,243 (540)     0.8684 (0.0005)  0.5004 (0.0012)  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "    'Sample': [],\n",
    "    'Imputed TP': [],\n",
    "    'Imputed FN': [],\n",
    "    'Imputed FP': [],\n",
    "    'Imputed Precision': [],\n",
    "    'Imputed Recall': [],\n",
    "    'WEGS 4P2X TP': [],\n",
    "    'WEGS 4P2X FN': [],\n",
    "    'WEGS 4P2X FP': [],\n",
    "    'WEGS 4P2X Precision': [],\n",
    "    'WEGS 4P2X Recall': [],\n",
    "    'WEGS 8P5X TP': [],\n",
    "    'WEGS 8P5X FN': [],\n",
    "    'WEGS 8P5X FP': [],\n",
    "    'WEGS 8P5X Precision': [],\n",
    "    'WEGS 8P5X Recall': [],\n",
    "}\n",
    "\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    data['Sample'].append(hgid)\n",
    "    \n",
    "    df_temp = df_array_imputed_variants[df_array_imputed_variants.HGID == hgid]\n",
    "    \n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL')][['TP', 'FN', 'FP', 'Recall', 'Precision']]\n",
    "    \n",
    "    data['Imputed TP'].append(f'{df_temp.TP.values[0]:,}')\n",
    "    data['Imputed FP'].append(f'{df_temp.FP.values[0]:,}')\n",
    "    data['Imputed FN'].append(f'{df_temp.FN.values[0]:,}')\n",
    "    data['Imputed Precision'].append(f'{df_temp.Precision.values[0]:.4f}')\n",
    "    data['Imputed Recall'].append(f'{df_temp.Recall.values[0]:.4f}')\n",
    "\n",
    "    df_temp = df_wegs_variants[df_wegs_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 4) & (df_temp.WGS_depth == 2)]\n",
    "    data['WEGS 4P2X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 4P2X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 4P2X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 4P2X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 4P2X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    \n",
    "    df_temp = df_wegs_variants[df_wegs_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 8) & (df_temp.WGS_depth == 5)]\n",
    "    data['WEGS 8P5X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 8P5X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 8P5X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 8P5X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 8P5X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    \n",
    "pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70c657c2-2b85-4ed0-90bf-208e2926c8b0",
   "metadata": {},
   "source": [
    "### TABLE: Precision and recall rates of variants imputed using the TOPMed reference panel genome-wide (PART 2 - InDels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "62fc946c-9af5-4bfb-9522-1ecbc3923e8f",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Imputed TP</th>\n",
       "      <th>Imputed FN</th>\n",
       "      <th>Imputed FP</th>\n",
       "      <th>Imputed Precision</th>\n",
       "      <th>Imputed Recall</th>\n",
       "      <th>WEGS 4P2X TP</th>\n",
       "      <th>WEGS 4P2X FN</th>\n",
       "      <th>WEGS 4P2X FP</th>\n",
       "      <th>WEGS 4P2X Precision</th>\n",
       "      <th>WEGS 4P2X Recall</th>\n",
       "      <th>WEGS 8P5X TP</th>\n",
       "      <th>WEGS 8P5X FN</th>\n",
       "      <th>WEGS 8P5X FP</th>\n",
       "      <th>WEGS 8P5X Precision</th>\n",
       "      <th>WEGS 8P5X Recall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002</td>\n",
       "      <td>199,347</td>\n",
       "      <td>323,042</td>\n",
       "      <td>2,089</td>\n",
       "      <td>0.9896</td>\n",
       "      <td>0.3816</td>\n",
       "      <td>132,493 (1,957)</td>\n",
       "      <td>389,896 (1,957)</td>\n",
       "      <td>55,634 (632)</td>\n",
       "      <td>0.7043 (0.0008)</td>\n",
       "      <td>0.2536 (0.0037)</td>\n",
       "      <td>198,767 (351)</td>\n",
       "      <td>323,622 (351)</td>\n",
       "      <td>68,172 (123)</td>\n",
       "      <td>0.7447 (0.0003)</td>\n",
       "      <td>0.3805 (0.0007)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG003</td>\n",
       "      <td>195,885</td>\n",
       "      <td>304,908</td>\n",
       "      <td>1,795</td>\n",
       "      <td>0.9909</td>\n",
       "      <td>0.3911</td>\n",
       "      <td>154,695 (1,172)</td>\n",
       "      <td>346,098 (1,172)</td>\n",
       "      <td>63,485 (423)</td>\n",
       "      <td>0.7091 (0.0003)</td>\n",
       "      <td>0.3089 (0.0023)</td>\n",
       "      <td>232,882 (368)</td>\n",
       "      <td>267,911 (368)</td>\n",
       "      <td>72,598 (142)</td>\n",
       "      <td>0.7624 (0.0002)</td>\n",
       "      <td>0.4650 (0.0007)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG004</td>\n",
       "      <td>197,298</td>\n",
       "      <td>310,976</td>\n",
       "      <td>2,006</td>\n",
       "      <td>0.9899</td>\n",
       "      <td>0.3882</td>\n",
       "      <td>111,566 (788)</td>\n",
       "      <td>396,708 (788)</td>\n",
       "      <td>50,204 (374)</td>\n",
       "      <td>0.6897 (0.0004)</td>\n",
       "      <td>0.2195 (0.0015)</td>\n",
       "      <td>168,158 (651)</td>\n",
       "      <td>340,116 (651)</td>\n",
       "      <td>62,636 (264)</td>\n",
       "      <td>0.7287 (0.0002)</td>\n",
       "      <td>0.3308 (0.0013)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Imputed TP Imputed FN Imputed FP Imputed Precision Imputed Recall  \\\n",
       "0  HG002    199,347    323,042      2,089            0.9896         0.3816   \n",
       "1  HG003    195,885    304,908      1,795            0.9909         0.3911   \n",
       "2  HG004    197,298    310,976      2,006            0.9899         0.3882   \n",
       "\n",
       "      WEGS 4P2X TP     WEGS 4P2X FN  WEGS 4P2X FP WEGS 4P2X Precision  \\\n",
       "0  132,493 (1,957)  389,896 (1,957)  55,634 (632)     0.7043 (0.0008)   \n",
       "1  154,695 (1,172)  346,098 (1,172)  63,485 (423)     0.7091 (0.0003)   \n",
       "2    111,566 (788)    396,708 (788)  50,204 (374)     0.6897 (0.0004)   \n",
       "\n",
       "  WEGS 4P2X Recall   WEGS 8P5X TP   WEGS 8P5X FN  WEGS 8P5X FP  \\\n",
       "0  0.2536 (0.0037)  198,767 (351)  323,622 (351)  68,172 (123)   \n",
       "1  0.3089 (0.0023)  232,882 (368)  267,911 (368)  72,598 (142)   \n",
       "2  0.2195 (0.0015)  168,158 (651)  340,116 (651)  62,636 (264)   \n",
       "\n",
       "  WEGS 8P5X Precision WEGS 8P5X Recall  \n",
       "0     0.7447 (0.0003)  0.3805 (0.0007)  \n",
       "1     0.7624 (0.0002)  0.4650 (0.0007)  \n",
       "2     0.7287 (0.0002)  0.3308 (0.0013)  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "    'Sample': [],\n",
    "    'Imputed TP': [],\n",
    "    'Imputed FN': [],\n",
    "    'Imputed FP': [],\n",
    "    'Imputed Precision': [],\n",
    "    'Imputed Recall': [],\n",
    "    'WEGS 4P2X TP': [],\n",
    "    'WEGS 4P2X FN': [],\n",
    "    'WEGS 4P2X FP': [],\n",
    "    'WEGS 4P2X Precision': [],\n",
    "    'WEGS 4P2X Recall': [],\n",
    "    'WEGS 8P5X TP': [],\n",
    "    'WEGS 8P5X FN': [],\n",
    "    'WEGS 8P5X FP': [],\n",
    "    'WEGS 8P5X Precision': [],\n",
    "    'WEGS 8P5X Recall': [],\n",
    "}\n",
    "\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    data['Sample'].append(hgid)\n",
    "    \n",
    "    df_temp = df_array_imputed_variants[df_array_imputed_variants.HGID == hgid]\n",
    "    \n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL')][['TP', 'FN', 'FP', 'Recall', 'Precision']]\n",
    "    \n",
    "    data['Imputed TP'].append(f'{df_temp.TP.values[0]:,}')\n",
    "    data['Imputed FP'].append(f'{df_temp.FP.values[0]:,}')\n",
    "    data['Imputed FN'].append(f'{df_temp.FN.values[0]:,}')\n",
    "    data['Imputed Precision'].append(f'{df_temp.Precision.values[0]:.4f}')\n",
    "    data['Imputed Recall'].append(f'{df_temp.Recall.values[0]:.4f}')\n",
    "\n",
    "    df_temp = df_wegs_variants[df_wegs_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 4) & (df_temp.WGS_depth == 2)]\n",
    "    data['WEGS 4P2X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 4P2X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 4P2X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 4P2X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 4P2X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    \n",
    "    df_temp = df_wegs_variants[df_wegs_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 8) & (df_temp.WGS_depth == 5)]\n",
    "    data['WEGS 8P5X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 8P5X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 8P5X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 8P5X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 8P5X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    \n",
    "pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e384582-69e5-4a78-b127-ac61c078889b",
   "metadata": {},
   "source": [
    "### TABLE: Imputed variants, their allele frequencies, and overlap with true positive (TP) variants in WEGS outside WES target regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "22feeb64-0ad5-4446-a71b-95264f1d0f04",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Truth vs imputed</th>\n",
       "      <th>N variants</th>\n",
       "      <th>WEGS 4P2X</th>\n",
       "      <th>WEGS 8P5X</th>\n",
       "      <th>AF ASJ</th>\n",
       "      <th>AF TOPMed</th>\n",
       "      <th>AF fold-change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002</td>\n",
       "      <td>AC_SAME</td>\n",
       "      <td>3,258,732</td>\n",
       "      <td>38.44 (0.42)</td>\n",
       "      <td>55.16 (0.05)</td>\n",
       "      <td>0.460 (0.239-0.703)</td>\n",
       "      <td>0.442 (0.231-0.680)</td>\n",
       "      <td>1.042 (0.925-1.197)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG002</td>\n",
       "      <td>AC_LESS</td>\n",
       "      <td>15,233</td>\n",
       "      <td>41.85 (0.48)</td>\n",
       "      <td>59.31 (0.11)</td>\n",
       "      <td>0.234 (0.059-0.483)</td>\n",
       "      <td>0.215 (0.043-0.453)</td>\n",
       "      <td>1.136 (0.926-1.509)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG002</td>\n",
       "      <td>AC_GREATER</td>\n",
       "      <td>17,161</td>\n",
       "      <td>90.55 (0.05)</td>\n",
       "      <td>92.75 (0.02)</td>\n",
       "      <td>0.329 (0.147-0.536)</td>\n",
       "      <td>0.340 (0.157-0.539)</td>\n",
       "      <td>1.007 (0.840-1.209)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG003</td>\n",
       "      <td>AC_SAME</td>\n",
       "      <td>3,225,810</td>\n",
       "      <td>44.84 (0.28)</td>\n",
       "      <td>64.45 (0.07)</td>\n",
       "      <td>0.461 (0.239-0.704)</td>\n",
       "      <td>0.443 (0.231-0.681)</td>\n",
       "      <td>1.043 (0.927-1.198)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG003</td>\n",
       "      <td>AC_LESS</td>\n",
       "      <td>12,665</td>\n",
       "      <td>46.75 (0.34)</td>\n",
       "      <td>67.09 (0.08)</td>\n",
       "      <td>0.222 (0.051-0.496)</td>\n",
       "      <td>0.215 (0.033-0.458)</td>\n",
       "      <td>1.129 (0.930-1.537)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HG003</td>\n",
       "      <td>AC_GREATER</td>\n",
       "      <td>13,855</td>\n",
       "      <td>92.38 (0.04)</td>\n",
       "      <td>94.60 (0.03)</td>\n",
       "      <td>0.302 (0.134-0.477)</td>\n",
       "      <td>0.318 (0.145-0.501)</td>\n",
       "      <td>1.006 (0.822-1.210)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HG004</td>\n",
       "      <td>AC_SAME</td>\n",
       "      <td>3,247,624</td>\n",
       "      <td>33.65 (0.19)</td>\n",
       "      <td>49.22 (0.12)</td>\n",
       "      <td>0.457 (0.237-0.702)</td>\n",
       "      <td>0.440 (0.230-0.679)</td>\n",
       "      <td>1.041 (0.923-1.196)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HG004</td>\n",
       "      <td>AC_LESS</td>\n",
       "      <td>15,357</td>\n",
       "      <td>38.89 (0.30)</td>\n",
       "      <td>55.00 (0.12)</td>\n",
       "      <td>0.219 (0.056-0.465)</td>\n",
       "      <td>0.205 (0.042-0.440)</td>\n",
       "      <td>1.114 (0.915-1.565)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HG004</td>\n",
       "      <td>AC_GREATER</td>\n",
       "      <td>16,457</td>\n",
       "      <td>89.38 (0.05)</td>\n",
       "      <td>91.94 (0.02)</td>\n",
       "      <td>0.325 (0.142-0.534)</td>\n",
       "      <td>0.330 (0.153-0.539)</td>\n",
       "      <td>1.022 (0.841-1.239)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Truth vs imputed N variants     WEGS 4P2X     WEGS 8P5X  \\\n",
       "0  HG002          AC_SAME  3,258,732  38.44 (0.42)  55.16 (0.05)   \n",
       "1  HG002          AC_LESS     15,233  41.85 (0.48)  59.31 (0.11)   \n",
       "2  HG002       AC_GREATER     17,161  90.55 (0.05)  92.75 (0.02)   \n",
       "3  HG003          AC_SAME  3,225,810  44.84 (0.28)  64.45 (0.07)   \n",
       "4  HG003          AC_LESS     12,665  46.75 (0.34)  67.09 (0.08)   \n",
       "5  HG003       AC_GREATER     13,855  92.38 (0.04)  94.60 (0.03)   \n",
       "6  HG004          AC_SAME  3,247,624  33.65 (0.19)  49.22 (0.12)   \n",
       "7  HG004          AC_LESS     15,357  38.89 (0.30)  55.00 (0.12)   \n",
       "8  HG004       AC_GREATER     16,457  89.38 (0.05)  91.94 (0.02)   \n",
       "\n",
       "                AF ASJ            AF TOPMed       AF fold-change  \n",
       "0  0.460 (0.239-0.703)  0.442 (0.231-0.680)  1.042 (0.925-1.197)  \n",
       "1  0.234 (0.059-0.483)  0.215 (0.043-0.453)  1.136 (0.926-1.509)  \n",
       "2  0.329 (0.147-0.536)  0.340 (0.157-0.539)  1.007 (0.840-1.209)  \n",
       "3  0.461 (0.239-0.704)  0.443 (0.231-0.681)  1.043 (0.927-1.198)  \n",
       "4  0.222 (0.051-0.496)  0.215 (0.033-0.458)  1.129 (0.930-1.537)  \n",
       "5  0.302 (0.134-0.477)  0.318 (0.145-0.501)  1.006 (0.822-1.210)  \n",
       "6  0.457 (0.237-0.702)  0.440 (0.230-0.679)  1.041 (0.923-1.196)  \n",
       "7  0.219 (0.056-0.465)  0.205 (0.042-0.440)  1.114 (0.915-1.565)  \n",
       "8  0.325 (0.142-0.534)  0.330 (0.153-0.539)  1.022 (0.841-1.239)  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "categories = ['AC_SAME', 'AC_LESS', 'AC_GREATER']\n",
    "\n",
    "data = {\n",
    "    'Sample': [],\n",
    "    'Truth vs imputed': [],\n",
    "    'N variants': [],\n",
    "    'WEGS 4P2X': [],\n",
    "    'WEGS 8P5X': [],\n",
    "    'AF ASJ': [],\n",
    "    'AF TOPMed': [],\n",
    "    'AF fold-change': [],\n",
    "}\n",
    "\n",
    "for hgid, df in imputed_arrays.items():\n",
    "    for cat in categories:\n",
    "        data['Sample'].append(hgid)\n",
    "        \n",
    "        df_temp = df[(df.IN_TARGET != True) & (df[cat] == True)]\n",
    "        \n",
    "        wegs = [len(df_temp[~df_temp.Name.isin(names)]) / len(df_temp) * 100 for names in wegs_bad_4p2x[hgid]]\n",
    "        data['WEGS 4P2X'].append(f'{np.mean(wegs):,.2f} ({stats.sem(wegs):,.2f})')\n",
    "        \n",
    "        wegs = [len(df_temp[~df_temp.Name.isin(names)]) / len(df_temp) * 100 for names in wegs_bad_8p5x[hgid]]\n",
    "        data['WEGS 8P5X'].append(f'{np.mean(wegs):,.2f} ({stats.sem(wegs):,.2f})')\n",
    "\n",
    "        data['Truth vs imputed'].append(cat)\n",
    "        data['N variants'].append(f'{len(df_temp):,}')\n",
    "\n",
    "        median = np.median(df_temp.ASJ_AF)\n",
    "        q25, q75 = np.quantile(df_temp.ASJ_AF, [0.25, 0.75])\n",
    "        data['AF ASJ'].append(f'{median:.3f} ({q25:.3f}-{q75:.3f})')\n",
    "\n",
    "        median = np.median(df_temp.TOPMED_AF)\n",
    "        q25, q75 = np.quantile(df_temp.TOPMED_AF, [0.25, 0.75])\n",
    "        data['AF TOPMed'].append(f'{median:.3f} ({q25:.3f}-{q75:.3f})')\n",
    "\n",
    "        median = np.median(df_temp.FOLD_CHANGE_AF)\n",
    "        q25, q75 = np.quantile(df_temp.FOLD_CHANGE_AF, [0.25, 0.75])\n",
    "        data['AF fold-change'].append(f'{median:.3f} ({q25:.3f}-{q75:.3f})')\n",
    "        \n",
    "pd.DataFrame(data)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "654d2fee-f60d-4606-a462-71eb7994eba4",
   "metadata": {},
   "source": [
    "### EXTRA TABLE: Imputed variants, their allele frequencies, and overlap with true positive (TP) variants in WEGS in targets only."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0d50dfb0-3d4d-4f3b-9aae-fb39e1b1a4af",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Truth vs imputed</th>\n",
       "      <th>N variants 4P+2X</th>\n",
       "      <th>% variants 4P+2X</th>\n",
       "      <th>N variants 8P+5X</th>\n",
       "      <th>% variants 8P+5X</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002</td>\n",
       "      <td>AC_SAME</td>\n",
       "      <td>20387.2</td>\n",
       "      <td>99.6</td>\n",
       "      <td>20387.2</td>\n",
       "      <td>99.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG002</td>\n",
       "      <td>AC_LESS</td>\n",
       "      <td>117.8</td>\n",
       "      <td>98.9</td>\n",
       "      <td>117.8</td>\n",
       "      <td>98.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG002</td>\n",
       "      <td>AC_GREATER</td>\n",
       "      <td>134.0</td>\n",
       "      <td>99.3</td>\n",
       "      <td>134.0</td>\n",
       "      <td>99.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG003</td>\n",
       "      <td>AC_SAME</td>\n",
       "      <td>20263.2</td>\n",
       "      <td>99.6</td>\n",
       "      <td>20263.2</td>\n",
       "      <td>99.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG003</td>\n",
       "      <td>AC_LESS</td>\n",
       "      <td>82.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HG003</td>\n",
       "      <td>AC_GREATER</td>\n",
       "      <td>83.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HG004</td>\n",
       "      <td>AC_SAME</td>\n",
       "      <td>20169.2</td>\n",
       "      <td>99.7</td>\n",
       "      <td>20169.2</td>\n",
       "      <td>99.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HG004</td>\n",
       "      <td>AC_LESS</td>\n",
       "      <td>183.0</td>\n",
       "      <td>98.9</td>\n",
       "      <td>183.0</td>\n",
       "      <td>98.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HG004</td>\n",
       "      <td>AC_GREATER</td>\n",
       "      <td>119.2</td>\n",
       "      <td>96.2</td>\n",
       "      <td>119.2</td>\n",
       "      <td>96.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Truth vs imputed N variants 4P+2X % variants 4P+2X N variants 8P+5X  \\\n",
       "0  HG002          AC_SAME          20387.2             99.6          20387.2   \n",
       "1  HG002          AC_LESS            117.8             98.9            117.8   \n",
       "2  HG002       AC_GREATER            134.0             99.3            134.0   \n",
       "3  HG003          AC_SAME          20263.2             99.6          20263.2   \n",
       "4  HG003          AC_LESS             82.0            100.0             82.0   \n",
       "5  HG003       AC_GREATER             83.0            100.0             83.0   \n",
       "6  HG004          AC_SAME          20169.2             99.7          20169.2   \n",
       "7  HG004          AC_LESS            183.0             98.9            183.0   \n",
       "8  HG004       AC_GREATER            119.2             96.2            119.2   \n",
       "\n",
       "  % variants 8P+5X  \n",
       "0             99.6  \n",
       "1             98.9  \n",
       "2             99.3  \n",
       "3             99.6  \n",
       "4            100.0  \n",
       "5            100.0  \n",
       "6             99.7  \n",
       "7             98.9  \n",
       "8             96.2  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "    'Sample': [],\n",
    "    'Truth vs imputed': [],\n",
    "    'N variants 4P+2X': [],\n",
    "    '% variants 4P+2X': [],\n",
    "    'N variants 8P+5X': [],\n",
    "    '% variants 8P+5X': [],\n",
    "}\n",
    "\n",
    "categories = ['AC_SAME', 'AC_LESS', 'AC_GREATER']\n",
    "\n",
    "for hgid, df in imputed_arrays.items():\n",
    "    for cat in categories:\n",
    "        data['Sample'].append(hgid)\n",
    "        if cat == 'All':\n",
    "            df_temp = df[(df.IN_TARGET == True)]\n",
    "        else:\n",
    "            df_temp = df[(df.IN_TARGET == True) & (df[cat] == True)]\n",
    "            \n",
    "        data['Truth vs imputed'].append(cat)\n",
    "            \n",
    "        n = []\n",
    "        perc = []\n",
    "        for wegs_bad in wegs_bad_4p2x[hgid]:\n",
    "            df_temp_subset = df_temp[(~df_temp.Name.isin(wegs_bad))]\n",
    "            n.append(len(df_temp_subset))\n",
    "            perc.append(len(df_temp_subset) / len(df_temp) * 100)\n",
    "        data['N variants 4P+2X'].append(f'{np.mean(n):.1f}')\n",
    "        data['% variants 4P+2X'].append(f'{np.mean(perc):.1f}')\n",
    "        \n",
    "        n = []\n",
    "        perc = []\n",
    "        for wegs_bad in wegs_bad_4p2x[hgid]:\n",
    "            df_temp_subset = df_temp[(~df_temp.Name.isin(wegs_bad))]\n",
    "            n.append(len(df_temp_subset))\n",
    "            perc.append(len(df_temp_subset) / len(df_temp) * 100)\n",
    "        data['N variants 8P+5X'].append(f'{np.mean(n):.1f}')\n",
    "        data['% variants 8P+5X'].append(f'{np.mean(perc):.1f}')\n",
    "        \n",
    "        \n",
    "pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7f4e936-6574-44a7-82a9-bc8b48f2c56b",
   "metadata": {},
   "source": [
    "### EXTRA TABLE: Target region SNVs missed by imputation but capruted by WEGS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "85a2a00d-a399-4a42-af2a-4324f12e60f2",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "### THE INPUT FILES FOR THE TABLE WERE GENERATED USING THIS CODE:\n",
    "\n",
    "# ### 1. Load True Positives in imputation, WEGS 4P2X, and WEGS 8P5X\n",
    "\n",
    "# imputed_arrays_target_tp = {}\n",
    "# for hgid, naid in hgid2naid.items():\n",
    "#     filename = f'{TOPMED_ARRAY_BENCHMARK_DIR}/Genomewide_with_AF/{hgid}_{naid}_{hgid2rel[hgid]}.dose.b37.happy_benchmark.vcf.gz'\n",
    "#     imputed_arrays_target_tp[hgid] = pd.DataFrame(read_tp_snvs(filename))\n",
    "    \n",
    "# wegs_target_tp_4p2x = dict()\n",
    "# for hgid, naid in hgid2naid.items():\n",
    "#     wegs_target_tp_4p2x[hgid] = []\n",
    "#     for filename in glob.glob(f'{WEGS_BENCHMARK_DIR}/4plex_2X_genomewide_with_AF/{hgid}_{naid}_*.norm.hard_filter.happy_benchmark.vcf.gz'):\n",
    "#         wegs_target_tp_4p2x[hgid].append(set(read_tp_snvs(filename)))\n",
    "\n",
    "# wegs_target_tp_8p5x = dict()\n",
    "# for hgid, naid in hgid2naid.items():\n",
    "#     wegs_target_tp_8p5x[hgid] = []\n",
    "#     for filename in glob.glob(f'{WEGS_BENCHMARK_DIR}/8plex_5X_genomewide_with_AF/{hgid}_{naid}_*.norm.hard_filter.happy_benchmark.vcf.gz'):\n",
    "#         wegs_target_tp_8p5x[hgid].append(set(read_tp_snvs(filename)))\n",
    "        \n",
    "# ### 2. Extract unique variants (which were True Positives only in WEGS)\n",
    "# for hgid, naid in hgid2naid.items():\n",
    "#     array_variants = imputed_arrays_target_tp[hgid]\n",
    "#     for i, wegs_variants in enumerate(wegs_target_tp_4p2x[hgid], 1):\n",
    "#         unique_to_array = array_variants.difference(wegs_variants)\n",
    "#         unique_to_wegs = wegs_variants.difference(array_variants)\n",
    "#         print(hgid, len(array_variants), len(wegs_variants), len(unique_to_array), len(unique_to_wegs))\n",
    "#         with open(f'{hgid}.{i}.4P2X.unique_to_wegs.txt', 'wt') as ofile:\n",
    "#             for name in sorted(unique_to_wegs):\n",
    "#                 chrom, pos, ref, alt = name.split(':')\n",
    "#                 ofile.write(f'{chrom} {pos} . {ref} {alt} . . .' + '\\n')\n",
    "                \n",
    "# for hgid, naid in hgid2naid.items():\n",
    "#     array_variants = imputed_arrays_target_tp[hgid]\n",
    "#     for i, wegs_variants in enumerate(wegs_target_tp_8p5x[hgid], 1):\n",
    "#         unique_to_array = array_variants.difference(wegs_variants)\n",
    "#         unique_to_wegs = wegs_variants.difference(array_variants)\n",
    "#         print(hgid, len(array_variants), len(wegs_variants), len(unique_to_array), len(unique_to_wegs))\n",
    "#         with open(f'{hgid}.{i}.8P5X.unique_to_wegs.txt', 'wt') as ofile:\n",
    "#             for name in sorted(unique_to_wegs):\n",
    "#                 chrom, pos, ref, alt = name.split(':')\n",
    "#                 ofile.write(f'{chrom} {pos} . {ref} {alt} . . .' + '\\n')\n",
    "\n",
    "# ### 3. RUN ONLINE VEP (with display most severe consequence option) on the *.unique_to_wegs.txt files, and save output as *.unique_to_wegs.vep.txt\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c1f92a59-1e9f-44ed-96a9-6a52fb24f7b1",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequencing</th>\n",
       "      <th>Sample</th>\n",
       "      <th>N_only_in_WEGS</th>\n",
       "      <th>N_nonsynonymous_only_in_WEGS</th>\n",
       "      <th>N_splice_only_in_WEGS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WEGS_4P2X</td>\n",
       "      <td>HG002</td>\n",
       "      <td>976</td>\n",
       "      <td>569</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WEGS_4P2X</td>\n",
       "      <td>HG002</td>\n",
       "      <td>970</td>\n",
       "      <td>559</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WEGS_4P2X</td>\n",
       "      <td>HG002</td>\n",
       "      <td>977</td>\n",
       "      <td>565</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WEGS_4P2X</td>\n",
       "      <td>HG002</td>\n",
       "      <td>977</td>\n",
       "      <td>565</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG002</td>\n",
       "      <td>981</td>\n",
       "      <td>569</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG002</td>\n",
       "      <td>967</td>\n",
       "      <td>558</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG002</td>\n",
       "      <td>986</td>\n",
       "      <td>571</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG002</td>\n",
       "      <td>970</td>\n",
       "      <td>558</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG002</td>\n",
       "      <td>981</td>\n",
       "      <td>567</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WEGS_4P2X</td>\n",
       "      <td>HG003</td>\n",
       "      <td>834</td>\n",
       "      <td>485</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>WEGS_4P2X</td>\n",
       "      <td>HG003</td>\n",
       "      <td>830</td>\n",
       "      <td>482</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>WEGS_4P2X</td>\n",
       "      <td>HG003</td>\n",
       "      <td>824</td>\n",
       "      <td>483</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>WEGS_4P2X</td>\n",
       "      <td>HG003</td>\n",
       "      <td>831</td>\n",
       "      <td>482</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG003</td>\n",
       "      <td>848</td>\n",
       "      <td>494</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG003</td>\n",
       "      <td>834</td>\n",
       "      <td>484</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG003</td>\n",
       "      <td>830</td>\n",
       "      <td>483</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG003</td>\n",
       "      <td>835</td>\n",
       "      <td>484</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG003</td>\n",
       "      <td>829</td>\n",
       "      <td>483</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG003</td>\n",
       "      <td>829</td>\n",
       "      <td>483</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>WEGS_4P2X</td>\n",
       "      <td>HG004</td>\n",
       "      <td>1008</td>\n",
       "      <td>554</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>WEGS_4P2X</td>\n",
       "      <td>HG004</td>\n",
       "      <td>1028</td>\n",
       "      <td>576</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>WEGS_4P2X</td>\n",
       "      <td>HG004</td>\n",
       "      <td>1000</td>\n",
       "      <td>556</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>WEGS_4P2X</td>\n",
       "      <td>HG004</td>\n",
       "      <td>1011</td>\n",
       "      <td>562</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG004</td>\n",
       "      <td>1003</td>\n",
       "      <td>557</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG004</td>\n",
       "      <td>1002</td>\n",
       "      <td>560</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG004</td>\n",
       "      <td>1002</td>\n",
       "      <td>552</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG004</td>\n",
       "      <td>1009</td>\n",
       "      <td>561</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>WEGS_8P5X</td>\n",
       "      <td>HG004</td>\n",
       "      <td>994</td>\n",
       "      <td>555</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sequencing Sample  N_only_in_WEGS  N_nonsynonymous_only_in_WEGS  \\\n",
       "0   WEGS_4P2X  HG002             976                           569   \n",
       "1   WEGS_4P2X  HG002             970                           559   \n",
       "2   WEGS_4P2X  HG002             977                           565   \n",
       "3   WEGS_4P2X  HG002             977                           565   \n",
       "4   WEGS_8P5X  HG002             981                           569   \n",
       "5   WEGS_8P5X  HG002             967                           558   \n",
       "6   WEGS_8P5X  HG002             986                           571   \n",
       "7   WEGS_8P5X  HG002             970                           558   \n",
       "8   WEGS_8P5X  HG002             981                           567   \n",
       "9   WEGS_4P2X  HG003             834                           485   \n",
       "10  WEGS_4P2X  HG003             830                           482   \n",
       "11  WEGS_4P2X  HG003             824                           483   \n",
       "12  WEGS_4P2X  HG003             831                           482   \n",
       "13  WEGS_8P5X  HG003             848                           494   \n",
       "14  WEGS_8P5X  HG003             834                           484   \n",
       "15  WEGS_8P5X  HG003             830                           483   \n",
       "16  WEGS_8P5X  HG003             835                           484   \n",
       "17  WEGS_8P5X  HG003             829                           483   \n",
       "18  WEGS_8P5X  HG003             829                           483   \n",
       "19  WEGS_4P2X  HG004            1008                           554   \n",
       "20  WEGS_4P2X  HG004            1028                           576   \n",
       "21  WEGS_4P2X  HG004            1000                           556   \n",
       "22  WEGS_4P2X  HG004            1011                           562   \n",
       "23  WEGS_8P5X  HG004            1003                           557   \n",
       "24  WEGS_8P5X  HG004            1002                           560   \n",
       "25  WEGS_8P5X  HG004            1002                           552   \n",
       "26  WEGS_8P5X  HG004            1009                           561   \n",
       "27  WEGS_8P5X  HG004             994                           555   \n",
       "\n",
       "    N_splice_only_in_WEGS  \n",
       "0                       3  \n",
       "1                       3  \n",
       "2                       3  \n",
       "3                       3  \n",
       "4                       3  \n",
       "5                       3  \n",
       "6                       3  \n",
       "7                       3  \n",
       "8                       3  \n",
       "9                       2  \n",
       "10                      2  \n",
       "11                      2  \n",
       "12                      2  \n",
       "13                      2  \n",
       "14                      2  \n",
       "15                      2  \n",
       "16                      2  \n",
       "17                      2  \n",
       "18                      2  \n",
       "19                      3  \n",
       "20                      3  \n",
       "21                      3  \n",
       "22                      3  \n",
       "23                      3  \n",
       "24                      3  \n",
       "25                      3  \n",
       "26                      3  \n",
       "27                      3  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = []\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    for wegs_name in ['4P2X', '8P5X']:\n",
    "        for filename in glob.glob(f'{TP_ONLY_IN_WEGS_DIR}/{hgid}.*.{wegs_name}.unique_to_wegs.vep.txt'):\n",
    "            df_temp = pd.read_table(filename)\n",
    "            total = len(df_temp)\n",
    "            n_nonsynonymous = len(df_temp[df_temp.Consequence.isin({'missense_variant', 'stop_gained', 'stop_lost', 'start_lost'})])\n",
    "            n_splice = len(df_temp[df_temp.Consequence.isin({'splice_acceptor_variant', 'splice_donor_variant'})])\n",
    "            data.append({\n",
    "                'Sequencing': f'WEGS_{wegs_name}',\n",
    "                'Sample': hgid,\n",
    "                'N_only_in_WEGS': total,\n",
    "                'N_nonsynonymous_only_in_WEGS': n_nonsynonymous,\n",
    "                'N_splice_only_in_WEGS': n_splice,\n",
    "            })\n",
    "\n",
    "pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10d090b1-1978-4457-81f9-45b032d4ae1c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
