{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cf9f5180-d798-4552-838b-e3ee66d1d863",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ignoring pip: markers 'python_version < \"3\"' don't match your environment\n",
      "Looking in links: /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo/avx2, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo/generic, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/generic\n",
      "Processing /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo/avx2/pysam-0.20.0+computecanada-cp39-cp39-linux_x86_64.whl\n",
      "Installing collected packages: pysam\n",
      "Successfully installed pysam\n",
      "Ignoring pip: markers 'python_version < \"3\"' don't match your environment\n",
      "Looking in links: /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo/avx2, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/gentoo/generic, /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/generic\n",
      "Processing /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/generic/intervaltree-3.1.0+computecanada-py2.py3-none-any.whl\n",
      "Processing /cvmfs/soft.computecanada.ca/custom/python/wheelhouse/generic/sortedcontainers-2.4.0+computecanada-py3-none-any.whl\n",
      "Installing collected packages: sortedcontainers, intervaltree\n",
      "Successfully installed intervaltree-3.1.0+computecanada sortedcontainers-2.4.0+computecanada\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!{sys.executable} -m pip install pysam --no-index\n",
    "!{sys.executable} -m pip install intervaltree --no-index\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4a32ff5c-98f2-478e-bbfc-42b151fc4e8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import stats\n",
    "import glob\n",
    "from intervaltree import Interval, IntervalTree\n",
    "import requests, sys"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec4bed38-36a4-437d-9608-65c30e55a358",
   "metadata": {},
   "source": [
    "# 1. Input files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "051f9bff-7cba-49bb-9d57-00edc75d373d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ARRAY_BENCHMARK_DIR = '/home/dtaliun/scratch/ExomePlus_paper/GenotypingArrays/'\n",
    "WEGS_BENCHMARK_DIR = '/home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal/'\n",
    "#WEGS_TOPMED_DIR = '/home/dtaliun/scratch/ExomePlus_paper/Analysis/WEGS_TOPMed_comparison/'\n",
    "GLIMPSE_BENCHMARK_DIR = '/home/dtaliun/scratch/ExomePlus_paper/Analysis/WEGS_imputation/Imputed_benchmark_genomewide'\n",
    "\n",
    "HAPPY_COLUMNS = ['Type', 'Filter', 'TRUTH.TP', 'QUERY.FP', 'TRUTH.FN', 'METRIC.Recall', 'METRIC.Precision', 'METRIC.F1_Score']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e782394a-a3ce-42b5-87fc-a36a5ec3dba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load sample annotations\n",
    "df_samples = pd.read_csv('../sample_annotations.txt', header = 0, sep = '\\t')\n",
    "\n",
    "hgid2naid = {\n",
    "    'HG002': 'NA24385',\n",
    "    'HG003': 'NA24149',\n",
    "    'HG004': 'NA24143'\n",
    "}\n",
    "\n",
    "hgid2rel = {\n",
    "    'HG002': 'Son',\n",
    "    'HG003': 'Father',\n",
    "    'HG004': 'Mother'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9bd6fe23-e2b0-4dd9-a853-2acdacda62be",
   "metadata": {},
   "outputs": [],
   "source": [
    "WEGS_TARGETS_BED = '/home/dtaliun/scratch/ExomePlus_paper/WES_targets/SureSelectHumanAllExonV7.b37.Target.bed'\n",
    "\n",
    "def load_targets_bed(filename):\n",
    "    targets = dict()\n",
    "    with open(filename, 'rt') as ifile:\n",
    "        for line in ifile:\n",
    "            chrom, start, stop = line.rstrip().split()\n",
    "            targets.setdefault(chrom, IntervalTree()).addi(int(start) + 1, int(stop) + 1) # because BED is 0-based and IntervalTree stores half-open intervals.\n",
    "    return targets\n",
    "\n",
    "WEGS_TARGETS = load_targets_bed(WEGS_TARGETS_BED)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b09c7ae4-790b-421c-8041-3d956edddb4b",
   "metadata": {},
   "source": [
    "# 2. Comparing total numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "eca26d5c-e441-4460-932e-2b1ec61c6d07",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_array_imputed_variants = []\n",
    "for row_index, row in df_samples[['HGID', 'Family']].drop_duplicates().iterrows():\n",
    "    filepath = f'{ARRAY_BENCHMARK_DIR}/TOPMed_b37_benchmark_genomewide/results/{row.HGID}_{hgid2naid[row.HGID]}_{row.Family}.dose.b37.happy_benchmark.summary.csv'\n",
    "    df = pd.read_csv(filepath, usecols = HAPPY_COLUMNS).rename(columns = {\n",
    "        'METRIC.Recall': 'Recall', \n",
    "        'METRIC.Precision': 'Precision',\n",
    "        'METRIC.F1_Score': 'F1_Score',\n",
    "        'TRUTH.TP': 'TP',\n",
    "        'QUERY.FP': 'FP',\n",
    "        'TRUTH.FN': 'FN'\n",
    "    })\n",
    "    df['Total_called'] = df['TP'] + df['FP']\n",
    "    df['HGID'] = row.HGID\n",
    "    df_array_imputed_variants.append(df)\n",
    "df_array_imputed_variants = pd.concat(df_array_imputed_variants).reset_index(drop = True)\n",
    "\n",
    "df_array_imputed_target_variants = []\n",
    "for row_index, row in df_samples[['HGID', 'Family']].drop_duplicates().iterrows():\n",
    "    filepath = f'{ARRAY_BENCHMARK_DIR}/TOPMed_b37_benchmark/results/{row.HGID}_{hgid2naid[row.HGID]}_{row.Family}.dose.b37.happy_benchmark.summary.csv'\n",
    "    df = pd.read_csv(filepath, usecols = HAPPY_COLUMNS).rename(columns = {\n",
    "        'METRIC.Recall': 'Recall', \n",
    "        'METRIC.Precision': 'Precision',\n",
    "        'METRIC.F1_Score': 'F1_Score',\n",
    "        'TRUTH.TP': 'TP',\n",
    "        'QUERY.FP': 'FP',\n",
    "        'TRUTH.FN': 'FN'\n",
    "    })\n",
    "    df['Total_called'] = df['TP'] + df['FP']\n",
    "    df['HGID'] = row.HGID\n",
    "    df_array_imputed_target_variants.append(df)\n",
    "df_array_imputed_target_variants = pd.concat(df_array_imputed_target_variants).reset_index(drop = True)\n",
    "\n",
    "df_wegs_variants = []\n",
    "for row_index, row in df_samples.iterrows():\n",
    "    if row.Plexing == 1:\n",
    "        continue\n",
    "    for wgs_depth in [2, 5]:\n",
    "        filepath = f'{WEGS_BENCHMARK_DIR}/{row.Plexing}plex_{wgs_depth}X_benchmark_genomewide/results/{row.Filename}.norm.hard_filter.happy_benchmark.summary.csv'\n",
    "        df = pd.read_csv(filepath, usecols = HAPPY_COLUMNS).rename(columns = {\n",
    "            'METRIC.Recall': 'Recall', \n",
    "            'METRIC.Precision': 'Precision',\n",
    "            'METRIC.F1_Score': 'F1_Score',\n",
    "            'TRUTH.TP': 'TP',\n",
    "            'QUERY.FP': 'FP',\n",
    "            'TRUTH.FN': 'FN'\n",
    "        })\n",
    "        df['Total_called'] = df['TP'] + df['FP']\n",
    "        df['ID'] = row.Filename\n",
    "        df['HGID'] = row.HGID\n",
    "        df['Batch'] = row.Batch\n",
    "        df['Plexing'] = row.Plexing\n",
    "        df['WGS_depth'] = wgs_depth\n",
    "        df_wegs_variants.append(df)\n",
    "df_wegs_variants = pd.concat(df_wegs_variants).reset_index(drop = True)\n",
    "\n",
    "df_wegs_target_variants = []\n",
    "for row_index, row in df_samples.iterrows():\n",
    "    if row.Plexing == 1:\n",
    "        continue\n",
    "    for wgs_depth in [2, 5]:\n",
    "        filepath = f'{WEGS_BENCHMARK_DIR}/{row.Plexing}plex_{wgs_depth}X_benchmark/results/{row.Filename}.norm.hard_filter.happy_benchmark.summary.csv'\n",
    "        df = pd.read_csv(filepath, usecols = HAPPY_COLUMNS).rename(columns = {\n",
    "            'METRIC.Recall': 'Recall', \n",
    "            'METRIC.Precision': 'Precision',\n",
    "            'METRIC.F1_Score': 'F1_Score',\n",
    "            'TRUTH.TP': 'TP',\n",
    "            'QUERY.FP': 'FP',\n",
    "            'TRUTH.FN': 'FN'\n",
    "        })\n",
    "        df['Total_called'] = df['TP'] + df['FP']\n",
    "        df['ID'] = row.Filename\n",
    "        df['HGID'] = row.HGID\n",
    "        df['Batch'] = row.Batch\n",
    "        df['Plexing'] = row.Plexing\n",
    "        df['WGS_depth'] = wgs_depth\n",
    "        df_wegs_target_variants.append(df)\n",
    "df_wegs_target_variants = pd.concat(df_wegs_target_variants).reset_index(drop = True)\n",
    "\n",
    "# df_wegs_imputed_target_variants = []\n",
    "# for row_index, row in df_samples.iterrows():\n",
    "#     if row.Plexing == 1 or row.Plexing != 4:\n",
    "#         continue\n",
    "#     # for wgs_depth in [2, 5]:\n",
    "#     for wgs_depth in [5]:\n",
    "#         filepath = f'{WEGS_GLIMPSE_BENCHMARK_DIR}/WEGS_{row.Plexing}plex_{wgs_depth}X/results/{row.Filename}.imputed.happy_benchmark.summary.csv'\n",
    "#         df = pd.read_csv(filepath, usecols = HAPPY_COLUMNS).rename(columns = {\n",
    "#             'METRIC.Recall': 'Recall', \n",
    "#             'METRIC.Precision': 'Precision',\n",
    "#             'METRIC.F1_Score': 'F1_Score',\n",
    "#             'TRUTH.TP': 'TP',\n",
    "#             'QUERY.FP': 'FP',\n",
    "#             'TRUTH.FN': 'FN'\n",
    "#         })\n",
    "#         df['Total_called'] = df['TP'] + df['FP']\n",
    "#         df['ID'] = row.Filename\n",
    "#         df['HGID'] = row.HGID\n",
    "#         df['Batch'] = row.Batch\n",
    "#         df['Plexing'] = row.Plexing\n",
    "#         df['WGS_depth'] = wgs_depth\n",
    "#         df_wegs_imputed_target_variants.append(df)\n",
    "# df_wegs_imputed_target_variants = pd.concat(df_wegs_imputed_target_variants).reset_index(drop = True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a0e965f-4f2c-4347-91e2-284e88032f77",
   "metadata": {},
   "source": [
    "## Inside targets (Supplementary Table 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4d5c3e36-0278-4c96-9ddb-c6df6ae80d34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG002 WEGS 4P2X 4\n",
      "HG002 WEGS 8P5X 5\n",
      "HG003 WEGS 4P2X 4\n",
      "HG003 WEGS 8P5X 6\n",
      "HG004 WEGS 4P2X 4\n",
      "HG004 WEGS 8P5X 5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Imputed TP</th>\n",
       "      <th>Imputed FN</th>\n",
       "      <th>Imputed FP</th>\n",
       "      <th>Imputed Precision</th>\n",
       "      <th>Imputed Recall</th>\n",
       "      <th>WEGS 4P2X TP</th>\n",
       "      <th>WEGS 4P2X FN</th>\n",
       "      <th>WEGS 4P2X FP</th>\n",
       "      <th>WEGS 4P2X Precision</th>\n",
       "      <th>WEGS 4P2X Recall</th>\n",
       "      <th>WEGS 8P5X TP</th>\n",
       "      <th>WEGS 8P5X FN</th>\n",
       "      <th>WEGS 8P5X FP</th>\n",
       "      <th>WEGS 8P5X Precision</th>\n",
       "      <th>WEGS 8P5X Recall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002</td>\n",
       "      <td>21,458</td>\n",
       "      <td>1,285</td>\n",
       "      <td>165</td>\n",
       "      <td>0.9924</td>\n",
       "      <td>0.9435</td>\n",
       "      <td>22,379 (7)</td>\n",
       "      <td>364 (7)</td>\n",
       "      <td>446 (6)</td>\n",
       "      <td>0.9805 (0.0002)</td>\n",
       "      <td>0.9840 (0.0003)</td>\n",
       "      <td>22,390 (5)</td>\n",
       "      <td>353 (5)</td>\n",
       "      <td>449 (7)</td>\n",
       "      <td>0.9803 (0.0003)</td>\n",
       "      <td>0.9845 (0.0002)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG003</td>\n",
       "      <td>21,477</td>\n",
       "      <td>1,112</td>\n",
       "      <td>111</td>\n",
       "      <td>0.9949</td>\n",
       "      <td>0.9508</td>\n",
       "      <td>22,231 (9)</td>\n",
       "      <td>358 (9)</td>\n",
       "      <td>425 (4)</td>\n",
       "      <td>0.9813 (0.0001)</td>\n",
       "      <td>0.9841 (0.0004)</td>\n",
       "      <td>22,249 (6)</td>\n",
       "      <td>340 (6)</td>\n",
       "      <td>406 (5)</td>\n",
       "      <td>0.9821 (0.0002)</td>\n",
       "      <td>0.9850 (0.0003)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG004</td>\n",
       "      <td>21,228</td>\n",
       "      <td>1,315</td>\n",
       "      <td>181</td>\n",
       "      <td>0.9915</td>\n",
       "      <td>0.9417</td>\n",
       "      <td>22,195 (9)</td>\n",
       "      <td>348 (9)</td>\n",
       "      <td>406 (7)</td>\n",
       "      <td>0.9820 (0.0003)</td>\n",
       "      <td>0.9846 (0.0004)</td>\n",
       "      <td>22,197 (5)</td>\n",
       "      <td>346 (5)</td>\n",
       "      <td>402 (3)</td>\n",
       "      <td>0.9822 (0.0001)</td>\n",
       "      <td>0.9847 (0.0002)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Imputed TP Imputed FN Imputed FP Imputed Precision Imputed Recall  \\\n",
       "0  HG002     21,458      1,285        165            0.9924         0.9435   \n",
       "1  HG003     21,477      1,112        111            0.9949         0.9508   \n",
       "2  HG004     21,228      1,315        181            0.9915         0.9417   \n",
       "\n",
       "  WEGS 4P2X TP WEGS 4P2X FN WEGS 4P2X FP WEGS 4P2X Precision WEGS 4P2X Recall  \\\n",
       "0   22,379 (7)      364 (7)      446 (6)     0.9805 (0.0002)  0.9840 (0.0003)   \n",
       "1   22,231 (9)      358 (9)      425 (4)     0.9813 (0.0001)  0.9841 (0.0004)   \n",
       "2   22,195 (9)      348 (9)      406 (7)     0.9820 (0.0003)  0.9846 (0.0004)   \n",
       "\n",
       "  WEGS 8P5X TP WEGS 8P5X FN WEGS 8P5X FP WEGS 8P5X Precision WEGS 8P5X Recall  \n",
       "0   22,390 (5)      353 (5)      449 (7)     0.9803 (0.0003)  0.9845 (0.0002)  \n",
       "1   22,249 (6)      340 (6)      406 (5)     0.9821 (0.0002)  0.9850 (0.0003)  \n",
       "2   22,197 (5)      346 (5)      402 (3)     0.9822 (0.0001)  0.9847 (0.0002)  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "    'Sample': [],\n",
    "    'Imputed TP': [],\n",
    "    'Imputed FN': [],\n",
    "    'Imputed FP': [],\n",
    "    'Imputed Precision': [],\n",
    "    'Imputed Recall': [],\n",
    "    'WEGS 4P2X TP': [],\n",
    "    'WEGS 4P2X FN': [],\n",
    "    'WEGS 4P2X FP': [],\n",
    "    'WEGS 4P2X Precision': [],\n",
    "    'WEGS 4P2X Recall': [],\n",
    "    'WEGS 8P5X TP': [],\n",
    "    'WEGS 8P5X FN': [],\n",
    "    'WEGS 8P5X FP': [],\n",
    "    'WEGS 8P5X Precision': [],\n",
    "    'WEGS 8P5X Recall': [],\n",
    "}\n",
    "\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    data['Sample'].append(hgid)\n",
    "    \n",
    "    df_temp = df_array_imputed_target_variants[df_array_imputed_target_variants.HGID == hgid]\n",
    "    \n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL')][['TP', 'FN', 'FP', 'Recall', 'Precision']]\n",
    "    \n",
    "    data['Imputed TP'].append(f'{df_temp.TP.values[0]:,}')\n",
    "    data['Imputed FP'].append(f'{df_temp.FP.values[0]:,}')\n",
    "    data['Imputed FN'].append(f'{df_temp.FN.values[0]:,}')\n",
    "    data['Imputed Precision'].append(f'{df_temp.Precision.values[0]:.4f}')\n",
    "    data['Imputed Recall'].append(f'{df_temp.Recall.values[0]:.4f}')\n",
    "\n",
    "    df_temp = df_wegs_target_variants[df_wegs_target_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 4) & (df_temp.WGS_depth == 2)]\n",
    "    data['WEGS 4P2X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 4P2X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 4P2X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 4P2X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 4P2X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    print(hgid, 'WEGS 4P2X', len(df_temp))\n",
    "    \n",
    "    df_temp = df_wegs_target_variants[df_wegs_target_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 8) & (df_temp.WGS_depth == 5)]\n",
    "    data['WEGS 8P5X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 8P5X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 8P5X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 8P5X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 8P5X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    print(hgid, 'WEGS 8P5X', len(df_temp))\n",
    "    \n",
    "pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "240e7740-7ad7-4712-9d3b-24f66036e7c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Imputed TP</th>\n",
       "      <th>Imputed FN</th>\n",
       "      <th>Imputed FP</th>\n",
       "      <th>Imputed Precision</th>\n",
       "      <th>Imputed Recall</th>\n",
       "      <th>WEGS 4P2X TP</th>\n",
       "      <th>WEGS 4P2X FN</th>\n",
       "      <th>WEGS 4P2X FP</th>\n",
       "      <th>WEGS 4P2X Precision</th>\n",
       "      <th>WEGS 4P2X Recall</th>\n",
       "      <th>WEGS 8P5X TP</th>\n",
       "      <th>WEGS 8P5X FN</th>\n",
       "      <th>WEGS 8P5X FP</th>\n",
       "      <th>WEGS 8P5X Precision</th>\n",
       "      <th>WEGS 8P5X Recall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002</td>\n",
       "      <td>271</td>\n",
       "      <td>411</td>\n",
       "      <td>8</td>\n",
       "      <td>0.9713</td>\n",
       "      <td>0.3974</td>\n",
       "      <td>648 (3)</td>\n",
       "      <td>34 (3)</td>\n",
       "      <td>57 (1)</td>\n",
       "      <td>0.9197 (0.0017)</td>\n",
       "      <td>0.9498 (0.0050)</td>\n",
       "      <td>647 (2)</td>\n",
       "      <td>35 (2)</td>\n",
       "      <td>58 (3)</td>\n",
       "      <td>0.9177 (0.0038)</td>\n",
       "      <td>0.9481 (0.0028)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG003</td>\n",
       "      <td>257</td>\n",
       "      <td>433</td>\n",
       "      <td>8</td>\n",
       "      <td>0.9698</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>649 (2)</td>\n",
       "      <td>41 (2)</td>\n",
       "      <td>52 (2)</td>\n",
       "      <td>0.9254 (0.0019)</td>\n",
       "      <td>0.9409 (0.0027)</td>\n",
       "      <td>651 (1)</td>\n",
       "      <td>39 (1)</td>\n",
       "      <td>52 (1)</td>\n",
       "      <td>0.9265 (0.0018)</td>\n",
       "      <td>0.9432 (0.0019)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG004</td>\n",
       "      <td>259</td>\n",
       "      <td>384</td>\n",
       "      <td>7</td>\n",
       "      <td>0.9738</td>\n",
       "      <td>0.4028</td>\n",
       "      <td>616 (2)</td>\n",
       "      <td>28 (2)</td>\n",
       "      <td>43 (2)</td>\n",
       "      <td>0.9357 (0.0035)</td>\n",
       "      <td>0.9572 (0.0026)</td>\n",
       "      <td>615 (2)</td>\n",
       "      <td>28 (2)</td>\n",
       "      <td>40 (2)</td>\n",
       "      <td>0.9393 (0.0033)</td>\n",
       "      <td>0.9568 (0.0031)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Imputed TP Imputed FN Imputed FP Imputed Precision Imputed Recall  \\\n",
       "0  HG002        271        411          8            0.9713         0.3974   \n",
       "1  HG003        257        433          8            0.9698         0.3725   \n",
       "2  HG004        259        384          7            0.9738         0.4028   \n",
       "\n",
       "  WEGS 4P2X TP WEGS 4P2X FN WEGS 4P2X FP WEGS 4P2X Precision WEGS 4P2X Recall  \\\n",
       "0      648 (3)       34 (3)       57 (1)     0.9197 (0.0017)  0.9498 (0.0050)   \n",
       "1      649 (2)       41 (2)       52 (2)     0.9254 (0.0019)  0.9409 (0.0027)   \n",
       "2      616 (2)       28 (2)       43 (2)     0.9357 (0.0035)  0.9572 (0.0026)   \n",
       "\n",
       "  WEGS 8P5X TP WEGS 8P5X FN WEGS 8P5X FP WEGS 8P5X Precision WEGS 8P5X Recall  \n",
       "0      647 (2)       35 (2)       58 (3)     0.9177 (0.0038)  0.9481 (0.0028)  \n",
       "1      651 (1)       39 (1)       52 (1)     0.9265 (0.0018)  0.9432 (0.0019)  \n",
       "2      615 (2)       28 (2)       40 (2)     0.9393 (0.0033)  0.9568 (0.0031)  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "    'Sample': [],\n",
    "    'Imputed TP': [],\n",
    "    'Imputed FN': [],\n",
    "    'Imputed FP': [],\n",
    "    'Imputed Precision': [],\n",
    "    'Imputed Recall': [],\n",
    "    'WEGS 4P2X TP': [],\n",
    "    'WEGS 4P2X FN': [],\n",
    "    'WEGS 4P2X FP': [],\n",
    "    'WEGS 4P2X Precision': [],\n",
    "    'WEGS 4P2X Recall': [],\n",
    "    'WEGS 8P5X TP': [],\n",
    "    'WEGS 8P5X FN': [],\n",
    "    'WEGS 8P5X FP': [],\n",
    "    'WEGS 8P5X Precision': [],\n",
    "    'WEGS 8P5X Recall': [],\n",
    "}\n",
    "\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    data['Sample'].append(hgid)\n",
    "    \n",
    "    df_temp = df_array_imputed_target_variants[df_array_imputed_target_variants.HGID == hgid]\n",
    "    \n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL')][['TP', 'FN', 'FP', 'Recall', 'Precision']]\n",
    "    \n",
    "    data['Imputed TP'].append(f'{df_temp.TP.values[0]:,}')\n",
    "    data['Imputed FP'].append(f'{df_temp.FP.values[0]:,}')\n",
    "    data['Imputed FN'].append(f'{df_temp.FN.values[0]:,}')\n",
    "    data['Imputed Precision'].append(f'{df_temp.Precision.values[0]:.4f}')\n",
    "    data['Imputed Recall'].append(f'{df_temp.Recall.values[0]:.4f}')\n",
    "\n",
    "    df_temp = df_wegs_target_variants[df_wegs_target_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 4) & (df_temp.WGS_depth == 2)]\n",
    "    data['WEGS 4P2X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 4P2X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 4P2X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 4P2X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 4P2X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    \n",
    "    df_temp = df_wegs_target_variants[df_wegs_target_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 8) & (df_temp.WGS_depth == 5)]\n",
    "    data['WEGS 8P5X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 8P5X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 8P5X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 8P5X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 8P5X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    \n",
    "pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4529249b-5cb6-4e06-bc6f-e4676e06f0b0",
   "metadata": {},
   "source": [
    "## Genome-wide (Supplementary Table 9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0644311a-2f6a-4cdd-8525-991b325b55ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG002 WEGS 4P2X 4\n",
      "HG002 WEGS 8P5X 5\n",
      "HG003 WEGS 4P2X 4\n",
      "HG003 WEGS 8P5X 6\n",
      "HG004 WEGS 4P2X 4\n",
      "HG004 WEGS 8P5X 5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Imputed TP</th>\n",
       "      <th>Imputed FN</th>\n",
       "      <th>Imputed FP</th>\n",
       "      <th>Imputed Precision</th>\n",
       "      <th>Imputed Recall</th>\n",
       "      <th>WEGS 4P2X TP</th>\n",
       "      <th>WEGS 4P2X FN</th>\n",
       "      <th>WEGS 4P2X FP</th>\n",
       "      <th>WEGS 4P2X Precision</th>\n",
       "      <th>WEGS 4P2X Recall</th>\n",
       "      <th>WEGS 8P5X TP</th>\n",
       "      <th>WEGS 8P5X FN</th>\n",
       "      <th>WEGS 8P5X FP</th>\n",
       "      <th>WEGS 8P5X Precision</th>\n",
       "      <th>WEGS 8P5X Recall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002</td>\n",
       "      <td>3,212,631</td>\n",
       "      <td>140,048</td>\n",
       "      <td>21,239</td>\n",
       "      <td>0.9934</td>\n",
       "      <td>0.9582</td>\n",
       "      <td>1,323,307 (13,949)</td>\n",
       "      <td>2,029,372 (13,949)</td>\n",
       "      <td>293,787 (1,558)</td>\n",
       "      <td>0.8183 (0.0023)</td>\n",
       "      <td>0.3947 (0.0042)</td>\n",
       "      <td>1,873,500 (1,770)</td>\n",
       "      <td>1,479,179 (1,770)</td>\n",
       "      <td>242,833 (353)</td>\n",
       "      <td>0.8853 (0.0002)</td>\n",
       "      <td>0.5588 (0.0005)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG003</td>\n",
       "      <td>3,181,715</td>\n",
       "      <td>132,428</td>\n",
       "      <td>17,060</td>\n",
       "      <td>0.9947</td>\n",
       "      <td>0.9600</td>\n",
       "      <td>1,518,733 (9,194)</td>\n",
       "      <td>1,795,410 (9,194)</td>\n",
       "      <td>325,169 (886)</td>\n",
       "      <td>0.8236 (0.0013)</td>\n",
       "      <td>0.4583 (0.0028)</td>\n",
       "      <td>2,151,757 (2,243)</td>\n",
       "      <td>1,162,386 (2,243)</td>\n",
       "      <td>228,413 (270)</td>\n",
       "      <td>0.9040 (0.0002)</td>\n",
       "      <td>0.6493 (0.0007)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG004</td>\n",
       "      <td>3,202,593</td>\n",
       "      <td>137,878</td>\n",
       "      <td>20,523</td>\n",
       "      <td>0.9936</td>\n",
       "      <td>0.9587</td>\n",
       "      <td>1,159,478 (6,227)</td>\n",
       "      <td>2,180,992 (6,227)</td>\n",
       "      <td>291,399 (438)</td>\n",
       "      <td>0.7991 (0.0011)</td>\n",
       "      <td>0.3471 (0.0019)</td>\n",
       "      <td>1,671,569 (3,941)</td>\n",
       "      <td>1,668,902 (3,941)</td>\n",
       "      <td>253,243 (540)</td>\n",
       "      <td>0.8684 (0.0005)</td>\n",
       "      <td>0.5004 (0.0012)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Imputed TP Imputed FN Imputed FP Imputed Precision Imputed Recall  \\\n",
       "0  HG002  3,212,631    140,048     21,239            0.9934         0.9582   \n",
       "1  HG003  3,181,715    132,428     17,060            0.9947         0.9600   \n",
       "2  HG004  3,202,593    137,878     20,523            0.9936         0.9587   \n",
       "\n",
       "         WEGS 4P2X TP        WEGS 4P2X FN     WEGS 4P2X FP  \\\n",
       "0  1,323,307 (13,949)  2,029,372 (13,949)  293,787 (1,558)   \n",
       "1   1,518,733 (9,194)   1,795,410 (9,194)    325,169 (886)   \n",
       "2   1,159,478 (6,227)   2,180,992 (6,227)    291,399 (438)   \n",
       "\n",
       "  WEGS 4P2X Precision WEGS 4P2X Recall       WEGS 8P5X TP       WEGS 8P5X FN  \\\n",
       "0     0.8183 (0.0023)  0.3947 (0.0042)  1,873,500 (1,770)  1,479,179 (1,770)   \n",
       "1     0.8236 (0.0013)  0.4583 (0.0028)  2,151,757 (2,243)  1,162,386 (2,243)   \n",
       "2     0.7991 (0.0011)  0.3471 (0.0019)  1,671,569 (3,941)  1,668,902 (3,941)   \n",
       "\n",
       "    WEGS 8P5X FP WEGS 8P5X Precision WEGS 8P5X Recall  \n",
       "0  242,833 (353)     0.8853 (0.0002)  0.5588 (0.0005)  \n",
       "1  228,413 (270)     0.9040 (0.0002)  0.6493 (0.0007)  \n",
       "2  253,243 (540)     0.8684 (0.0005)  0.5004 (0.0012)  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "    'Sample': [],\n",
    "    'Imputed TP': [],\n",
    "    'Imputed FN': [],\n",
    "    'Imputed FP': [],\n",
    "    'Imputed Precision': [],\n",
    "    'Imputed Recall': [],\n",
    "    'WEGS 4P2X TP': [],\n",
    "    'WEGS 4P2X FN': [],\n",
    "    'WEGS 4P2X FP': [],\n",
    "    'WEGS 4P2X Precision': [],\n",
    "    'WEGS 4P2X Recall': [],\n",
    "    'WEGS 8P5X TP': [],\n",
    "    'WEGS 8P5X FN': [],\n",
    "    'WEGS 8P5X FP': [],\n",
    "    'WEGS 8P5X Precision': [],\n",
    "    'WEGS 8P5X Recall': [],\n",
    "}\n",
    "\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    data['Sample'].append(hgid)\n",
    "    \n",
    "    df_temp = df_array_imputed_variants[df_array_imputed_variants.HGID == hgid]\n",
    "    \n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL')][['TP', 'FN', 'FP', 'Recall', 'Precision']]\n",
    "    \n",
    "    data['Imputed TP'].append(f'{df_temp.TP.values[0]:,}')\n",
    "    data['Imputed FP'].append(f'{df_temp.FP.values[0]:,}')\n",
    "    data['Imputed FN'].append(f'{df_temp.FN.values[0]:,}')\n",
    "    data['Imputed Precision'].append(f'{df_temp.Precision.values[0]:.4f}')\n",
    "    data['Imputed Recall'].append(f'{df_temp.Recall.values[0]:.4f}')\n",
    "\n",
    "    df_temp = df_wegs_variants[df_wegs_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 4) & (df_temp.WGS_depth == 2)]\n",
    "    data['WEGS 4P2X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 4P2X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 4P2X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 4P2X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 4P2X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    print(hgid, 'WEGS 4P2X', len(df_temp))\n",
    "    \n",
    "    df_temp = df_wegs_variants[df_wegs_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 8) & (df_temp.WGS_depth == 5)]\n",
    "    data['WEGS 8P5X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 8P5X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 8P5X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 8P5X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 8P5X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    print(hgid, 'WEGS 8P5X', len(df_temp))\n",
    "    \n",
    "pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "62fc946c-9af5-4bfb-9522-1ecbc3923e8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG002 WEGS 4P2X 4\n",
      "HG002 WEGS 8P5X 5\n",
      "HG003 WEGS 4P2X 4\n",
      "HG003 WEGS 8P5X 6\n",
      "HG004 WEGS 4P2X 4\n",
      "HG004 WEGS 8P5X 5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Imputed TP</th>\n",
       "      <th>Imputed FN</th>\n",
       "      <th>Imputed FP</th>\n",
       "      <th>Imputed Precision</th>\n",
       "      <th>Imputed Recall</th>\n",
       "      <th>WEGS 4P2X TP</th>\n",
       "      <th>WEGS 4P2X FN</th>\n",
       "      <th>WEGS 4P2X FP</th>\n",
       "      <th>WEGS 4P2X Precision</th>\n",
       "      <th>WEGS 4P2X Recall</th>\n",
       "      <th>WEGS 8P5X TP</th>\n",
       "      <th>WEGS 8P5X FN</th>\n",
       "      <th>WEGS 8P5X FP</th>\n",
       "      <th>WEGS 8P5X Precision</th>\n",
       "      <th>WEGS 8P5X Recall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002</td>\n",
       "      <td>199,347</td>\n",
       "      <td>323,042</td>\n",
       "      <td>2,089</td>\n",
       "      <td>0.9896</td>\n",
       "      <td>0.3816</td>\n",
       "      <td>132,493 (1,957)</td>\n",
       "      <td>389,896 (1,957)</td>\n",
       "      <td>55,634 (632)</td>\n",
       "      <td>0.7043 (0.0008)</td>\n",
       "      <td>0.2536 (0.0037)</td>\n",
       "      <td>198,767 (351)</td>\n",
       "      <td>323,622 (351)</td>\n",
       "      <td>68,172 (123)</td>\n",
       "      <td>0.7447 (0.0003)</td>\n",
       "      <td>0.3805 (0.0007)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG003</td>\n",
       "      <td>195,885</td>\n",
       "      <td>304,908</td>\n",
       "      <td>1,795</td>\n",
       "      <td>0.9909</td>\n",
       "      <td>0.3911</td>\n",
       "      <td>154,695 (1,172)</td>\n",
       "      <td>346,098 (1,172)</td>\n",
       "      <td>63,485 (423)</td>\n",
       "      <td>0.7091 (0.0003)</td>\n",
       "      <td>0.3089 (0.0023)</td>\n",
       "      <td>232,882 (368)</td>\n",
       "      <td>267,911 (368)</td>\n",
       "      <td>72,598 (142)</td>\n",
       "      <td>0.7624 (0.0002)</td>\n",
       "      <td>0.4650 (0.0007)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG004</td>\n",
       "      <td>197,298</td>\n",
       "      <td>310,976</td>\n",
       "      <td>2,006</td>\n",
       "      <td>0.9899</td>\n",
       "      <td>0.3882</td>\n",
       "      <td>111,566 (788)</td>\n",
       "      <td>396,708 (788)</td>\n",
       "      <td>50,204 (374)</td>\n",
       "      <td>0.6897 (0.0004)</td>\n",
       "      <td>0.2195 (0.0015)</td>\n",
       "      <td>168,158 (651)</td>\n",
       "      <td>340,116 (651)</td>\n",
       "      <td>62,636 (264)</td>\n",
       "      <td>0.7287 (0.0002)</td>\n",
       "      <td>0.3308 (0.0013)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Imputed TP Imputed FN Imputed FP Imputed Precision Imputed Recall  \\\n",
       "0  HG002    199,347    323,042      2,089            0.9896         0.3816   \n",
       "1  HG003    195,885    304,908      1,795            0.9909         0.3911   \n",
       "2  HG004    197,298    310,976      2,006            0.9899         0.3882   \n",
       "\n",
       "      WEGS 4P2X TP     WEGS 4P2X FN  WEGS 4P2X FP WEGS 4P2X Precision  \\\n",
       "0  132,493 (1,957)  389,896 (1,957)  55,634 (632)     0.7043 (0.0008)   \n",
       "1  154,695 (1,172)  346,098 (1,172)  63,485 (423)     0.7091 (0.0003)   \n",
       "2    111,566 (788)    396,708 (788)  50,204 (374)     0.6897 (0.0004)   \n",
       "\n",
       "  WEGS 4P2X Recall   WEGS 8P5X TP   WEGS 8P5X FN  WEGS 8P5X FP  \\\n",
       "0  0.2536 (0.0037)  198,767 (351)  323,622 (351)  68,172 (123)   \n",
       "1  0.3089 (0.0023)  232,882 (368)  267,911 (368)  72,598 (142)   \n",
       "2  0.2195 (0.0015)  168,158 (651)  340,116 (651)  62,636 (264)   \n",
       "\n",
       "  WEGS 8P5X Precision WEGS 8P5X Recall  \n",
       "0     0.7447 (0.0003)  0.3805 (0.0007)  \n",
       "1     0.7624 (0.0002)  0.4650 (0.0007)  \n",
       "2     0.7287 (0.0002)  0.3308 (0.0013)  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "    'Sample': [],\n",
    "    'Imputed TP': [],\n",
    "    'Imputed FN': [],\n",
    "    'Imputed FP': [],\n",
    "    'Imputed Precision': [],\n",
    "    'Imputed Recall': [],\n",
    "    'WEGS 4P2X TP': [],\n",
    "    'WEGS 4P2X FN': [],\n",
    "    'WEGS 4P2X FP': [],\n",
    "    'WEGS 4P2X Precision': [],\n",
    "    'WEGS 4P2X Recall': [],\n",
    "    'WEGS 8P5X TP': [],\n",
    "    'WEGS 8P5X FN': [],\n",
    "    'WEGS 8P5X FP': [],\n",
    "    'WEGS 8P5X Precision': [],\n",
    "    'WEGS 8P5X Recall': [],\n",
    "}\n",
    "\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    data['Sample'].append(hgid)\n",
    "    \n",
    "    df_temp = df_array_imputed_variants[df_array_imputed_variants.HGID == hgid]\n",
    "    \n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL')][['TP', 'FN', 'FP', 'Recall', 'Precision']]\n",
    "    \n",
    "    data['Imputed TP'].append(f'{df_temp.TP.values[0]:,}')\n",
    "    data['Imputed FP'].append(f'{df_temp.FP.values[0]:,}')\n",
    "    data['Imputed FN'].append(f'{df_temp.FN.values[0]:,}')\n",
    "    data['Imputed Precision'].append(f'{df_temp.Precision.values[0]:.4f}')\n",
    "    data['Imputed Recall'].append(f'{df_temp.Recall.values[0]:.4f}')\n",
    "\n",
    "    df_temp = df_wegs_variants[df_wegs_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 4) & (df_temp.WGS_depth == 2)]\n",
    "    data['WEGS 4P2X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 4P2X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 4P2X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 4P2X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 4P2X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    print(hgid, 'WEGS 4P2X', len(df_temp))\n",
    "    \n",
    "    df_temp = df_wegs_variants[df_wegs_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 8) & (df_temp.WGS_depth == 5)]\n",
    "    data['WEGS 8P5X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['WEGS 8P5X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['WEGS 8P5X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['WEGS 8P5X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['WEGS 8P5X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    print(hgid, 'WEGS 8P5X', len(df_temp))\n",
    "    \n",
    "pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6688eb38-bde4-44f8-b816-3c5fe44cd178",
   "metadata": {},
   "source": [
    "# 3. Frequency comparison (Supplementary Table 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3814611b-ab1a-4ece-84d3-937faa84ce7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_af(filename):\n",
    "    with pysam.VariantFile(filename, 'r') as ifile:\n",
    "        for record in ifile.fetch():\n",
    "            if 'ASJ_AF' not in record.info or 'ASJ_AN' not in record.info:\n",
    "                continue\n",
    "            if 'TOMPED_AF' not in record.info:\n",
    "                continue\n",
    "            if len(record.alts) > 1: # skip multi-allelic\n",
    "                continue\n",
    "            if record.info['ASJ_AN'] < 1000: # keep only those with sufficient ASJ sample size in gnomad \n",
    "                continue\n",
    "            asj_af = record.info.get('ASJ_AF', [0.0])[0]\n",
    "            if (asj_af <= 0) or (asj_af >= 1.0): # remove those which are monomorphic in ASJ according to gnomad\n",
    "                continue\n",
    "            topmed_af = record.info.get('TOMPED_AF', [0.0])[0]\n",
    "            if (topmed_af <= 0) or (topmed_af >= 1.0): # remove those which are monomorphic in TOPMed\n",
    "                continue\n",
    "            if record.samples['QUERY']['BD'] == 'UNK' or record.samples['TRUTH']['BD'] == 'UNK': # variants outside the truth set regions\n",
    "                continue\n",
    "            if record.samples['TRUTH']['BD'] == 'N': # skip halfcalls such as \"./1\"\n",
    "                continue\n",
    "            targets = WEGS_TARGETS[record.chrom]\n",
    "            entry = {\n",
    "                     'Name': f'{record.chrom}:{record.pos}:{record.ref}:{record.alts[0]}',\n",
    "                     'ASJ_AF': asj_af,\n",
    "                     'TOPMED_AF': topmed_af,\n",
    "                     'FOLD_CHANGE_AF': asj_af / topmed_af,\n",
    "                     'IN_TARGET': targets.overlaps(record.start - 1) or targets.overlaps(record.stop - 1),\n",
    "                     'AC_LESS': False,\n",
    "                     'AC_GREATER': False,\n",
    "                     'AC_SAME': False\n",
    "                    }\n",
    "            if record.samples['TRUTH']['BD'] == '.':\n",
    "                truth_gt = 0\n",
    "            else:\n",
    "                truth_gt = sum(record.samples['TRUTH']['GT'])\n",
    "            if record.samples['QUERY']['BD'] == '.':\n",
    "                query_gt = 0\n",
    "            else:\n",
    "                query_gt = sum(record.samples['QUERY']['GT'])            \n",
    "            if truth_gt > query_gt:\n",
    "                entry['AC_LESS'] = True\n",
    "            elif truth_gt < query_gt:\n",
    "                entry['AC_GREATER'] = True\n",
    "            else:\n",
    "                entry['AC_SAME'] = True\n",
    "            yield(entry)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7ee735bc-59c9-473a-bf3f-6837d7b2a869",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_bad_variant_calls(filename):\n",
    "    with pysam.VariantFile(filename, 'r') as ifile:\n",
    "        for record in ifile.fetch():\n",
    "            if 'ASJ_AF' not in record.info or 'ASJ_AN' not in record.info:\n",
    "                continue\n",
    "            if 'TOMPED_AF' not in record.info:\n",
    "                continue\n",
    "            if len(record.alts) > 1: # skip multi-allelic\n",
    "                continue\n",
    "            if record.info['ASJ_AN'] < 1000: # keep only those with sufficient ASJ sample size in gnomad \n",
    "                continue\n",
    "            asj_af = record.info.get('ASJ_AF', [0.0])[0]\n",
    "            if (asj_af <= 0) or (asj_af >= 1.0): # remove those which are monomorphic in ASJ according to gnomad\n",
    "                continue\n",
    "            topmed_af = record.info.get('TOMPED_AF', [0.0])[0]\n",
    "            if (topmed_af <= 0) or (topmed_af >= 1.0): # remove those which are monomorphic in TOPMed\n",
    "                continue\n",
    "            if record.samples['QUERY']['BD'] == 'UNK' or record.samples['TRUTH']['BD'] == 'UNK': # variants outside the truth set regions\n",
    "                continue\n",
    "            if record.samples['TRUTH']['BD'] == 'N': # skip halfcalls such as \"./1\"\n",
    "                continue\n",
    "            if record.samples['TRUTH']['BD'] == '.':\n",
    "                truth_gt = 0\n",
    "            else:\n",
    "                truth_gt = sum(record.samples['TRUTH']['GT'])\n",
    "            if record.samples['QUERY']['BD'] == '.':\n",
    "                query_gt = 0\n",
    "            else:\n",
    "                query_gt = sum(record.samples['QUERY']['GT'])            \n",
    "            if truth_gt != query_gt:\n",
    "                yield f'{record.chrom}:{record.pos}:{record.ref}:{record.alts[0]}'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c5632161-1116-4214-9761-71ae08fdc8e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n"
     ]
    }
   ],
   "source": [
    "imputed_arrays = {}\n",
    "for hdid, naid in hgid2naid.items():\n",
    "    filename = f'{ARRAY_BENCHMARK_DIR}/TOPMed_b37_benchmark_genomewide/results_topmed_af/{hdid}_{naid}_{hgid2rel[hdid]}.dose.b37.happy_benchmark.vcf.gz'\n",
    "    imputed_arrays[hdid] = pd.DataFrame(read_af(filename))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a2cdeda7-07fb-491d-b1dc-03e951ba1feb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n"
     ]
    }
   ],
   "source": [
    "wegs_4p2x = dict()\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    wegs_4p2x[hgid] = []\n",
    "    for filename in glob.glob(f'{WEGS_BENCHMARK_DIR}/4plex_2X_benchmark_genomewide/results_topmed_af/{hgid}_{naid}_*.norm.hard_filter.happy_benchmark.vcf.gz'):\n",
    "        bad_calls = set(read_bad_variant_calls(filename))\n",
    "        wegs_4p2x[hgid].append(bad_calls)\n",
    "\n",
    "wegs_8p5x = dict()\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    wegs_8p5x[hgid] = []\n",
    "    for filename in glob.glob(f'{WEGS_BENCHMARK_DIR}/8plex_5X_benchmark_genomewide/results_topmed_af/{hgid}_{naid}_*.norm.hard_filter.happy_benchmark.vcf.gz'):\n",
    "        bad_calls = set(read_bad_variant_calls(filename))\n",
    "        wegs_8p5x[hgid].append(bad_calls)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "22feeb64-0ad5-4446-a71b-95264f1d0f04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Truth vs imputed</th>\n",
       "      <th>N variants</th>\n",
       "      <th>WEGS 4P2X</th>\n",
       "      <th>WEGS 8P5X</th>\n",
       "      <th>AF ASJ</th>\n",
       "      <th>AF TOPMed</th>\n",
       "      <th>AF fold-change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002</td>\n",
       "      <td>AC_SAME</td>\n",
       "      <td>3,258,732</td>\n",
       "      <td>38.44 (0.42)</td>\n",
       "      <td>55.16 (0.05)</td>\n",
       "      <td>0.460 (0.239-0.703)</td>\n",
       "      <td>0.442 (0.231-0.680)</td>\n",
       "      <td>1.042 (0.925-1.197)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG002</td>\n",
       "      <td>AC_LESS</td>\n",
       "      <td>15,233</td>\n",
       "      <td>41.85 (0.48)</td>\n",
       "      <td>59.31 (0.11)</td>\n",
       "      <td>0.234 (0.059-0.483)</td>\n",
       "      <td>0.215 (0.043-0.453)</td>\n",
       "      <td>1.136 (0.926-1.509)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG002</td>\n",
       "      <td>AC_GREATER</td>\n",
       "      <td>17,161</td>\n",
       "      <td>90.55 (0.05)</td>\n",
       "      <td>92.75 (0.02)</td>\n",
       "      <td>0.329 (0.147-0.536)</td>\n",
       "      <td>0.340 (0.157-0.539)</td>\n",
       "      <td>1.007 (0.840-1.209)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG003</td>\n",
       "      <td>AC_SAME</td>\n",
       "      <td>3,225,810</td>\n",
       "      <td>44.84 (0.28)</td>\n",
       "      <td>64.45 (0.07)</td>\n",
       "      <td>0.461 (0.239-0.704)</td>\n",
       "      <td>0.443 (0.231-0.681)</td>\n",
       "      <td>1.043 (0.927-1.198)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG003</td>\n",
       "      <td>AC_LESS</td>\n",
       "      <td>12,665</td>\n",
       "      <td>46.75 (0.34)</td>\n",
       "      <td>67.09 (0.08)</td>\n",
       "      <td>0.222 (0.051-0.496)</td>\n",
       "      <td>0.215 (0.033-0.458)</td>\n",
       "      <td>1.129 (0.930-1.537)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HG003</td>\n",
       "      <td>AC_GREATER</td>\n",
       "      <td>13,855</td>\n",
       "      <td>92.38 (0.04)</td>\n",
       "      <td>94.60 (0.03)</td>\n",
       "      <td>0.302 (0.134-0.477)</td>\n",
       "      <td>0.318 (0.145-0.501)</td>\n",
       "      <td>1.006 (0.822-1.210)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HG004</td>\n",
       "      <td>AC_SAME</td>\n",
       "      <td>3,247,624</td>\n",
       "      <td>33.65 (0.19)</td>\n",
       "      <td>49.22 (0.12)</td>\n",
       "      <td>0.457 (0.237-0.702)</td>\n",
       "      <td>0.440 (0.230-0.679)</td>\n",
       "      <td>1.041 (0.923-1.196)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HG004</td>\n",
       "      <td>AC_LESS</td>\n",
       "      <td>15,357</td>\n",
       "      <td>38.89 (0.30)</td>\n",
       "      <td>55.00 (0.12)</td>\n",
       "      <td>0.219 (0.056-0.465)</td>\n",
       "      <td>0.205 (0.042-0.440)</td>\n",
       "      <td>1.114 (0.915-1.565)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HG004</td>\n",
       "      <td>AC_GREATER</td>\n",
       "      <td>16,457</td>\n",
       "      <td>89.38 (0.05)</td>\n",
       "      <td>91.94 (0.02)</td>\n",
       "      <td>0.325 (0.142-0.534)</td>\n",
       "      <td>0.330 (0.153-0.539)</td>\n",
       "      <td>1.022 (0.841-1.239)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Truth vs imputed N variants     WEGS 4P2X     WEGS 8P5X  \\\n",
       "0  HG002          AC_SAME  3,258,732  38.44 (0.42)  55.16 (0.05)   \n",
       "1  HG002          AC_LESS     15,233  41.85 (0.48)  59.31 (0.11)   \n",
       "2  HG002       AC_GREATER     17,161  90.55 (0.05)  92.75 (0.02)   \n",
       "3  HG003          AC_SAME  3,225,810  44.84 (0.28)  64.45 (0.07)   \n",
       "4  HG003          AC_LESS     12,665  46.75 (0.34)  67.09 (0.08)   \n",
       "5  HG003       AC_GREATER     13,855  92.38 (0.04)  94.60 (0.03)   \n",
       "6  HG004          AC_SAME  3,247,624  33.65 (0.19)  49.22 (0.12)   \n",
       "7  HG004          AC_LESS     15,357  38.89 (0.30)  55.00 (0.12)   \n",
       "8  HG004       AC_GREATER     16,457  89.38 (0.05)  91.94 (0.02)   \n",
       "\n",
       "                AF ASJ            AF TOPMed       AF fold-change  \n",
       "0  0.460 (0.239-0.703)  0.442 (0.231-0.680)  1.042 (0.925-1.197)  \n",
       "1  0.234 (0.059-0.483)  0.215 (0.043-0.453)  1.136 (0.926-1.509)  \n",
       "2  0.329 (0.147-0.536)  0.340 (0.157-0.539)  1.007 (0.840-1.209)  \n",
       "3  0.461 (0.239-0.704)  0.443 (0.231-0.681)  1.043 (0.927-1.198)  \n",
       "4  0.222 (0.051-0.496)  0.215 (0.033-0.458)  1.129 (0.930-1.537)  \n",
       "5  0.302 (0.134-0.477)  0.318 (0.145-0.501)  1.006 (0.822-1.210)  \n",
       "6  0.457 (0.237-0.702)  0.440 (0.230-0.679)  1.041 (0.923-1.196)  \n",
       "7  0.219 (0.056-0.465)  0.205 (0.042-0.440)  1.114 (0.915-1.565)  \n",
       "8  0.325 (0.142-0.534)  0.330 (0.153-0.539)  1.022 (0.841-1.239)  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "categories = ['AC_SAME', 'AC_LESS', 'AC_GREATER']\n",
    "\n",
    "data = {\n",
    "    'Sample': [],\n",
    "    'Truth vs imputed': [],\n",
    "    'N variants': [],\n",
    "    'WEGS 4P2X': [],\n",
    "    'WEGS 8P5X': [],\n",
    "    'AF ASJ': [],\n",
    "    'AF TOPMed': [],\n",
    "    'AF fold-change': [],\n",
    "}\n",
    "\n",
    "for hdid, df in imputed_arrays.items():\n",
    "    for cat in categories:\n",
    "        data['Sample'].append(hdid)\n",
    "        \n",
    "        df_temp = df[(df.IN_TARGET != True) & (df[cat] == True)]\n",
    "        \n",
    "        wegs = [len(df_temp[~df_temp.Name.isin(names)]) / len(df_temp) * 100 for names in wegs_4p2x[hdid]]\n",
    "        data['WEGS 4P2X'].append(f'{np.mean(wegs):,.2f} ({stats.sem(wegs):,.2f})')\n",
    "        \n",
    "        wegs = [len(df_temp[~df_temp.Name.isin(names)]) / len(df_temp) * 100 for names in wegs_8p5x[hdid]]\n",
    "        data['WEGS 8P5X'].append(f'{np.mean(wegs):,.2f} ({stats.sem(wegs):,.2f})')\n",
    "\n",
    "        data['Truth vs imputed'].append(cat)\n",
    "        data['N variants'].append(f'{len(df_temp):,}')\n",
    "\n",
    "        median = np.median(df_temp.ASJ_AF)\n",
    "        q25, q75 = np.quantile(df_temp.ASJ_AF, [0.25, 0.75])\n",
    "        data['AF ASJ'].append(f'{median:.3f} ({q25:.3f}-{q75:.3f})')\n",
    "\n",
    "        median = np.median(df_temp.TOPMED_AF)\n",
    "        q25, q75 = np.quantile(df_temp.TOPMED_AF, [0.25, 0.75])\n",
    "        data['AF TOPMed'].append(f'{median:.3f} ({q25:.3f}-{q75:.3f})')\n",
    "\n",
    "        median = np.median(df_temp.FOLD_CHANGE_AF)\n",
    "        q25, q75 = np.quantile(df_temp.FOLD_CHANGE_AF, [0.25, 0.75])\n",
    "        data['AF fold-change'].append(f'{median:.3f} ({q25:.3f}-{q75:.3f})')\n",
    "        \n",
    "pd.DataFrame(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0d50dfb0-3d4d-4f3b-9aae-fb39e1b1a4af",
   "metadata": {},
   "outputs": [],
   "source": [
    "## SAME AS ABOVE, BUT WITHIN THE TARGET REGIONS\n",
    "\n",
    "# data = {\n",
    "#     'Sample': [],\n",
    "#     'Truth vs imputed': [],\n",
    "#     'N variants 4P+2X': [],\n",
    "#     '% variants 4P+2X': [],\n",
    "#     'N variants 8P+5X': [],\n",
    "#     '% variants 8P+5X': [],\n",
    "# }\n",
    "\n",
    "# for hdid, df in imputed_arrays.items():\n",
    "#     for cat in categories:\n",
    "#         data['Sample'].append(hdid)\n",
    "#         if cat == 'All':\n",
    "#             df_temp = df[(df.IN_TARGET == True)]\n",
    "#         else:\n",
    "#             df_temp = df[(df.IN_TARGET == True) & (df[cat] == True)]\n",
    "            \n",
    "#         data['Truth vs imputed'].append(cat)\n",
    "            \n",
    "#         n = []\n",
    "#         perc = []\n",
    "#         for wegs_bad in wegs_4p2x[hgid]:\n",
    "#             df_temp_subset = df_temp[(~df_temp.Name.isin(wegs_bad))]\n",
    "#             n.append(len(df_temp_subset))\n",
    "#             perc.append(len(df_temp_subset) / len(df_temp) * 100)\n",
    "#         data['N variants 4P+2X'].append(f'{np.mean(n):.1f}')\n",
    "#         data['% variants 4P+2X'].append(f'{np.mean(perc):.1f}')\n",
    "        \n",
    "#         n = []\n",
    "#         perc = []\n",
    "#         for wegs_bad in wegs_8p5x[hgid]:\n",
    "#             df_temp_subset = df_temp[(~df_temp.Name.isin(wegs_bad))]\n",
    "#             n.append(len(df_temp_subset))\n",
    "#             perc.append(len(df_temp_subset) / len(df_temp) * 100)\n",
    "#         data['N variants 8P+5X'].append(f'{np.mean(n):.1f}')\n",
    "#         data['% variants 8P+5X'].append(f'{np.mean(perc):.1f}')\n",
    "        \n",
    "        \n",
    "# pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e13297eb-9c76-4baf-b388-03a49743cffa",
   "metadata": {},
   "source": [
    "# 4. Annotations of target region SNVs missed by imputation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "76587594-8c6b-4654-90a7-aa65ca212065",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_tp_snvs(filename):\n",
    "    with pysam.VariantFile(filename, 'r') as ifile:\n",
    "        for record in ifile.fetch():\n",
    "            if len(record.alts) > 1: # skip multi-allelic\n",
    "                continue\n",
    "            if len(record.ref) != 1 or len(record.alts[0]) != 1: # keep only SNVs\n",
    "                continue\n",
    "            if record.samples['QUERY']['BD'] == 'UNK' or record.samples['TRUTH']['BD'] == 'UNK': # variants outside the truth set regions\n",
    "                continue\n",
    "            if record.samples['TRUTH']['BD'] == 'N': # skip halfcalls such as \"./1\"\n",
    "                continue\n",
    "            targets = WEGS_TARGETS[record.chrom]\n",
    "            if not (targets.overlaps(record.start - 1) or targets.overlaps(record.stop - 1)): # use only variants within target regions\n",
    "                continue\n",
    "            if record.samples['TRUTH']['BD'] == '.':\n",
    "                truth_gt = 0\n",
    "            else:\n",
    "                truth_gt = sum(record.samples['TRUTH']['GT'])\n",
    "            if record.samples['QUERY']['BD'] == '.':\n",
    "                query_gt = 0\n",
    "            else:\n",
    "                query_gt = sum(record.samples['QUERY']['GT'])            \n",
    "            if truth_gt != query_gt: # skip not TP\n",
    "                continue\n",
    "            yield f'{record.chrom}:{record.pos}:{record.ref}:{record.alts[0]}'\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f37c6288-648d-47ad-b8c6-a7e65d8c4970",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/GenotypingArrays//TOPMed_b37_benchmark_genomewide/results/HG002_NA24385_Son.dose.b37.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/GenotypingArrays//TOPMed_b37_benchmark_genomewide/results/HG003_NA24149_Father.dose.b37.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/GenotypingArrays//TOPMed_b37_benchmark_genomewide/results/HG004_NA24143_Mother.dose.b37.happy_benchmark.vcf.gz.tbi\n"
     ]
    }
   ],
   "source": [
    "imputed_arrays_target_tp = {}\n",
    "for hdid, naid in hgid2naid.items():\n",
    "    # filename = f'/TOPMed_b37_benchmark_genomewide/results/{row.HGID}_{hgid2naid[row.HGID]}_{row.Family}.dose.b37.happy_benchmark.summary.csv'\n",
    "    filename = f'{ARRAY_BENCHMARK_DIR}/TOPMed_b37_benchmark_genomewide/results/{hdid}_{naid}_{hgid2rel[hdid]}.dose.b37.happy_benchmark.vcf.gz'\n",
    "    imputed_arrays_target_tp[hdid] = set(read_tp_snvs(filename))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c7bf4cd6-0560-4551-83af-795e51f5df7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//4plex_2X_benchmark_genomewide/results/HG002_NA24385_Son_SSXTHSV27054-SSXTHSV25054.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//4plex_2X_benchmark_genomewide/results/HG002_NA24385_Son_NewIndex_SSXTHSV27061-SSXTHSV25061.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//4plex_2X_benchmark_genomewide/results/HG002_NA24385_Son_SSXTHSV27058-SSXTHSV25058.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//4plex_2X_benchmark_genomewide/results/HG002_NA24385_Son_SSXTHSV27050-SSXTHSV25050.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//4plex_2X_benchmark_genomewide/results/HG003_NA24149_Father_SSXTHSV27060-SSXTHSV25060.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//4plex_2X_benchmark_genomewide/results/HG003_NA24149_Father_SSXTHSV27056-SSXTHSV25056.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//4plex_2X_benchmark_genomewide/results/HG003_NA24149_Father_NewIndex_SSXTHSV27053-SSXTHSV25053.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//4plex_2X_benchmark_genomewide/results/HG003_NA24149_Father_SSXTHSV27052-SSXTHSV25052.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//4plex_2X_benchmark_genomewide/results/HG004_NA24143_Mother_SSXTHSV27051-SSXTHSV25051.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//4plex_2X_benchmark_genomewide/results/HG004_NA24143_Mother_NewIndex_SSXTHSV27057-SSXTHSV25057.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//4plex_2X_benchmark_genomewide/results/HG004_NA24143_Mother_SSXTHSV27059-SSXTHSV25059.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//4plex_2X_benchmark_genomewide/results/HG004_NA24143_Mother_SSXTHSV27055-SSXTHSV25055.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n"
     ]
    }
   ],
   "source": [
    "wegs_target_tp_4p2x = dict()\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    wegs_target_tp_4p2x[hgid] = []\n",
    "    for filename in glob.glob(f'{WEGS_BENCHMARK_DIR}/4plex_2X_benchmark_genomewide/results/{hgid}_{naid}_*.norm.hard_filter.happy_benchmark.vcf.gz'):\n",
    "        wegs_target_tp_4p2x[hgid].append(set(read_tp_snvs(filename)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "eed13b89-ee57-4208-857e-0654724d2c9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG002 21144 22038 83 977\n",
      "HG002 21144 22026 94 976\n",
      "HG002 21144 22048 73 977\n",
      "HG002 21144 22054 60 970\n",
      "HG003 21035 21767 99 831\n",
      "HG003 21035 21800 69 834\n",
      "HG003 21035 21779 80 824\n",
      "HG003 21035 21799 66 830\n",
      "HG004 20935 21886 60 1011\n",
      "HG004 20935 21895 68 1028\n",
      "HG004 20935 21853 82 1000\n",
      "HG004 20935 21878 65 1008\n",
      "977 565 3\n",
      "976 569 3\n",
      "977 565 3\n",
      "970 559 3\n",
      "831 482 2\n",
      "834 485 2\n",
      "824 483 2\n",
      "830 482 2\n",
      "1011 562 3\n",
      "1028 576 3\n",
      "1000 556 3\n",
      "1008 554 3\n"
     ]
    }
   ],
   "source": [
    "## 1) EXTRACT UNIQUE VARIANTS TO ANNOTATE USING VEP\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    array_variants = imputed_arrays_target_tp[hgid]\n",
    "    for i, wegs_variants in enumerate(wegs_target_tp_4p2x[hgid], 1):\n",
    "        unique_to_array = array_variants.difference(wegs_variants)\n",
    "        unique_to_wegs = wegs_variants.difference(array_variants)\n",
    "        print(hgid, len(array_variants), len(wegs_variants), len(unique_to_array), len(unique_to_wegs))\n",
    "        with open(f'{hgid}.{i}.4P2X.unique_to_wegs.txt', 'wt') as ofile:\n",
    "            for name in sorted(unique_to_wegs):\n",
    "                chrom, pos, ref, alt = name.split(':')\n",
    "                ofile.write(f'{chrom} {pos} . {ref} {alt} . . .' + '\\n')\n",
    "                \n",
    "## 2) RUN ONLINE VEP (with display most severe consequence option) on the *.unique_to_wegs.txt files, and save output as *.unique_to_wegs.vep.txt\n",
    "\n",
    "## 3) GET COUNTS\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    for i, wegs_variants in enumerate(wegs_target_tp_4p2x[hgid], 1):\n",
    "        df_temp = pd.read_table(f'{hgid}.{i}.4P2X.unique_to_wegs.vep.txt')\n",
    "        total = len(df_temp)\n",
    "        n_nonsynonymous = len(df_temp[df_temp.Consequence.isin({'missense_variant', 'stop_gained', 'stop_lost', 'start_lost'})])\n",
    "        n_splice = len(df_temp[df_temp.Consequence.isin({'splice_acceptor_variant', 'splice_donor_variant'})])\n",
    "        print(total, n_nonsynonymous, n_splice)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d8cb1f6d-ae8c-44e1-9d1a-616fb7aa46e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//8plex_5X_benchmark_genomewide/results/HG002_NA24385_Son_NewIndex_SSXTHSV27077-SSXTHSV25077.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//8plex_5X_benchmark_genomewide/results/HG002_NA24385_Son_SSXTHSV27062-SSXTHSV25062.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//8plex_5X_benchmark_genomewide/results/HG003_NA24149_Father_NewIndex_SSXTHSV27068-SSXTHSV25068.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//8plex_5X_benchmark_genomewide/results/HG003_NA24149_Father_SSXTHSV27074-SSXTHSV25074.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//8plex_5X_benchmark_genomewide/results/HG004_NA24143_Mother_SSXTHSV27064-SSXTHSV25064.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n",
      "[W::bcf_hdr_register_hrec] The definition of Flag \"INFO/IMPORT_FAIL\" is invalid, forcing Number=0\n",
      "[W::hts_idx_load3] The index file is older than the data file: /home/dtaliun/scratch/ExomePlus_paper/Multiplexing_V7/WEGS_BAMs_MarkDuplicates_Recal//8plex_5X_benchmark_genomewide/results/HG004_NA24143_Mother_SSXTHSV27072-SSXTHSV25072.norm.hard_filter.happy_benchmark.vcf.gz.tbi\n"
     ]
    }
   ],
   "source": [
    "wegs_target_tp_8p5x = dict()\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    wegs_target_tp_8p5x[hgid] = []\n",
    "    for filename in glob.glob(f'{WEGS_BENCHMARK_DIR}/8plex_5X_benchmark_genomewide/results/{hgid}_{naid}_*.norm.hard_filter.happy_benchmark.vcf.gz'):\n",
    "        wegs_target_tp_8p5x[hgid].append(set(read_tp_snvs(filename)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0b693423-2ebe-469f-ab34-092d0703f400",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG002 21144 22059 66 981\n",
      "HG002 21144 22049 76 981\n",
      "HG002 21144 22039 75 970\n",
      "HG002 21144 22040 71 967\n",
      "HG002 21144 22071 59 986\n",
      "HG003 21035 21800 64 829\n",
      "HG003 21035 21828 55 848\n",
      "HG003 21035 21801 63 829\n",
      "HG003 21035 21793 72 830\n",
      "HG003 21035 21797 73 835\n",
      "HG003 21035 21787 82 834\n",
      "HG004 20935 21898 46 1009\n",
      "HG004 20935 21881 56 1002\n",
      "HG004 20935 21873 64 1002\n",
      "HG004 20935 21888 50 1003\n",
      "HG004 20935 21867 62 994\n",
      "981 567 3\n",
      "981 569 3\n",
      "970 558 3\n",
      "967 558 3\n",
      "986 571 3\n",
      "829 483 2\n",
      "848 494 2\n",
      "829 483 2\n",
      "830 483 2\n",
      "835 484 2\n",
      "834 484 2\n",
      "1009 561 3\n",
      "1002 560 3\n",
      "1002 552 3\n",
      "1003 557 3\n",
      "994 555 3\n"
     ]
    }
   ],
   "source": [
    "## 1) EXTRACT UNIQUE VARIANTS TO ANNOTATE USING VEP\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    array_variants = imputed_arrays_target_tp[hgid]\n",
    "    for i, wegs_variants in enumerate(wegs_target_tp_8p5x[hgid], 1):\n",
    "        unique_to_array = array_variants.difference(wegs_variants)\n",
    "        unique_to_wegs = wegs_variants.difference(array_variants)\n",
    "        print(hgid, len(array_variants), len(wegs_variants), len(unique_to_array), len(unique_to_wegs))\n",
    "        with open(f'{hgid}.{i}.8P5X.unique_to_wegs.txt', 'wt') as ofile:\n",
    "            for name in sorted(unique_to_wegs):\n",
    "                chrom, pos, ref, alt = name.split(':')\n",
    "                ofile.write(f'{chrom} {pos} . {ref} {alt} . . .' + '\\n')\n",
    "                \n",
    "## 2) RUN ONLINE VEP (with display most severe consequence option) on the *.unique_to_wegs.txt files, and save output as *.unique_to_wegs.vep.txt\n",
    "\n",
    "## 3) GET COUNTS\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    for i, wegs_variants in enumerate(wegs_target_tp_8p5x[hgid], 1):\n",
    "        df_temp = pd.read_table(f'{hgid}.{i}.8P5X.unique_to_wegs.vep.txt')\n",
    "        total = len(df_temp)\n",
    "        n_nonsynonymous = len(df_temp[df_temp.Consequence.isin({'missense_variant', 'stop_gained', 'stop_lost', 'start_lost'})])\n",
    "        n_splice = len(df_temp[df_temp.Consequence.isin({'splice_acceptor_variant', 'splice_donor_variant'})])\n",
    "        print(total, n_nonsynonymous, n_splice)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdf18fd4-3233-40ed-b1c5-7ca1a2d8e1eb",
   "metadata": {},
   "source": [
    "## 5. WEGS imputation with GLIMPSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8d220cd0-98ff-476c-a791-8430ff7b6c2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_glimpse_variants = []\n",
    "for row_index, row in df_samples.iterrows():\n",
    "    for plexing, wgs_depth in [(4, 2), (8, 5)]:\n",
    "        if row.Plexing != plexing:\n",
    "            continue\n",
    "    # if row.Plexing == 1:\n",
    "    #     continue\n",
    "    # for wgs_depth in [2, 5]:\n",
    "        filepath = f'{GLIMPSE_BENCHMARK_DIR}/WEGS_{row.Plexing}plex_{wgs_depth}X/results/{row.Filename}.imputed.merged_with_WEGS.happy_benchmark.summary.csv'\n",
    "        df = pd.read_csv(filepath, usecols = HAPPY_COLUMNS).rename(columns = {\n",
    "            'METRIC.Recall': 'Recall', \n",
    "            'METRIC.Precision': 'Precision',\n",
    "            'METRIC.F1_Score': 'F1_Score',\n",
    "            'TRUTH.TP': 'TP',\n",
    "            'QUERY.FP': 'FP',\n",
    "            'TRUTH.FN': 'FN'\n",
    "        })\n",
    "        df['Total_called'] = df['TP'] + df['FP']\n",
    "        df['ID'] = row.Filename\n",
    "        df['HGID'] = row.HGID\n",
    "        df['Batch'] = row.Batch\n",
    "        df['Plexing'] = row.Plexing\n",
    "        df['WGS_depth'] = wgs_depth\n",
    "        df_glimpse_variants.append(df)\n",
    "df_glimpse_variants = pd.concat(df_glimpse_variants).reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f7e2b67d-18fa-44a4-954c-7f0e65a99ffa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Imputed 4P2X TP</th>\n",
       "      <th>Imputed 4P2X FN</th>\n",
       "      <th>Imputed 4P2X FP</th>\n",
       "      <th>Imputed 4P2X Precision</th>\n",
       "      <th>Imputed 4P2X Recall</th>\n",
       "      <th>Imputed 8P5X TP</th>\n",
       "      <th>Imputed 8P5X FN</th>\n",
       "      <th>Imputed 8P5X FP</th>\n",
       "      <th>Imputed 8P5X Precision</th>\n",
       "      <th>Imputed 8P5X Recall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002</td>\n",
       "      <td>2,736,662 (2,636)</td>\n",
       "      <td>616,017 (2,636)</td>\n",
       "      <td>372,221 (1,713)</td>\n",
       "      <td>0.8803 (0.0006)</td>\n",
       "      <td>0.8163 (0.0008)</td>\n",
       "      <td>2,922,557 (375)</td>\n",
       "      <td>430,122 (375)</td>\n",
       "      <td>301,471 (328)</td>\n",
       "      <td>0.9065 (0.0001)</td>\n",
       "      <td>0.8717 (0.0001)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG003</td>\n",
       "      <td>2,709,032 (1,998)</td>\n",
       "      <td>605,111 (1,998)</td>\n",
       "      <td>388,043 (965)</td>\n",
       "      <td>0.8747 (0.0004)</td>\n",
       "      <td>0.8174 (0.0006)</td>\n",
       "      <td>2,919,299 (567)</td>\n",
       "      <td>394,844 (567)</td>\n",
       "      <td>267,597 (277)</td>\n",
       "      <td>0.9160 (0.0001)</td>\n",
       "      <td>0.8809 (0.0002)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG004</td>\n",
       "      <td>2,706,645 (1,212)</td>\n",
       "      <td>633,826 (1,212)</td>\n",
       "      <td>381,866 (504)</td>\n",
       "      <td>0.8764 (0.0002)</td>\n",
       "      <td>0.8103 (0.0004)</td>\n",
       "      <td>2,879,156 (815)</td>\n",
       "      <td>461,315 (815)</td>\n",
       "      <td>323,954 (586)</td>\n",
       "      <td>0.8989 (0.0002)</td>\n",
       "      <td>0.8619 (0.0002)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample    Imputed 4P2X TP  Imputed 4P2X FN  Imputed 4P2X FP  \\\n",
       "0  HG002  2,736,662 (2,636)  616,017 (2,636)  372,221 (1,713)   \n",
       "1  HG003  2,709,032 (1,998)  605,111 (1,998)    388,043 (965)   \n",
       "2  HG004  2,706,645 (1,212)  633,826 (1,212)    381,866 (504)   \n",
       "\n",
       "  Imputed 4P2X Precision Imputed 4P2X Recall  Imputed 8P5X TP Imputed 8P5X FN  \\\n",
       "0        0.8803 (0.0006)     0.8163 (0.0008)  2,922,557 (375)   430,122 (375)   \n",
       "1        0.8747 (0.0004)     0.8174 (0.0006)  2,919,299 (567)   394,844 (567)   \n",
       "2        0.8764 (0.0002)     0.8103 (0.0004)  2,879,156 (815)   461,315 (815)   \n",
       "\n",
       "  Imputed 8P5X FP Imputed 8P5X Precision Imputed 8P5X Recall  \n",
       "0   301,471 (328)        0.9065 (0.0001)     0.8717 (0.0001)  \n",
       "1   267,597 (277)        0.9160 (0.0001)     0.8809 (0.0002)  \n",
       "2   323,954 (586)        0.8989 (0.0002)     0.8619 (0.0002)  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "    'Sample': [],\n",
    "    'Imputed 4P2X TP': [],\n",
    "    'Imputed 4P2X FN': [],\n",
    "    'Imputed 4P2X FP': [],\n",
    "    'Imputed 4P2X Precision': [],\n",
    "    'Imputed 4P2X Recall': [],\n",
    "    'Imputed 8P5X TP': [],\n",
    "    'Imputed 8P5X FN': [],\n",
    "    'Imputed 8P5X FP': [],\n",
    "    'Imputed 8P5X Precision': [],\n",
    "    'Imputed 8P5X Recall': [],\n",
    "}\n",
    "\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    data['Sample'].append(hgid)\n",
    "    \n",
    "    df_temp = df_glimpse_variants[df_glimpse_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 4) & (df_temp.WGS_depth == 2)]\n",
    "    data['Imputed 4P2X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['Imputed 4P2X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['Imputed 4P2X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['Imputed 4P2X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['Imputed 4P2X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    # print(hgid, 'Imputed WEGS 4P2X', len(df_temp))\n",
    "    \n",
    "    df_temp = df_glimpse_variants[df_glimpse_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'SNP') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 8) & (df_temp.WGS_depth == 5)]\n",
    "    data['Imputed 8P5X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['Imputed 8P5X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['Imputed 8P5X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['Imputed 8P5X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['Imputed 8P5X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    # print(hgid, 'Imputed WEGS 8P5X', len(df_temp))\n",
    "    \n",
    "pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "8e3bfcdd-1806-4631-94c8-c3322e3ff3fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Imputed 4P2X TP</th>\n",
       "      <th>Imputed 4P2X FN</th>\n",
       "      <th>Imputed 4P2X FP</th>\n",
       "      <th>Imputed 4P2X Precision</th>\n",
       "      <th>Imputed 4P2X Recall</th>\n",
       "      <th>Imputed 8P5X TP</th>\n",
       "      <th>Imputed 8P5X FN</th>\n",
       "      <th>Imputed 8P5X FP</th>\n",
       "      <th>Imputed 8P5X Precision</th>\n",
       "      <th>Imputed 8P5X Recall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002</td>\n",
       "      <td>251,862 (1,203)</td>\n",
       "      <td>270,527 (1,203)</td>\n",
       "      <td>65,080 (614)</td>\n",
       "      <td>0.7947 (0.0008)</td>\n",
       "      <td>0.4821 (0.0023)</td>\n",
       "      <td>294,850 (254)</td>\n",
       "      <td>227,539 (254)</td>\n",
       "      <td>76,268 (112)</td>\n",
       "      <td>0.7945 (0.0002)</td>\n",
       "      <td>0.5644 (0.0005)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG003</td>\n",
       "      <td>256,262 (716)</td>\n",
       "      <td>244,530 (716)</td>\n",
       "      <td>70,962 (416)</td>\n",
       "      <td>0.7832 (0.0005)</td>\n",
       "      <td>0.5117 (0.0014)</td>\n",
       "      <td>305,742 (233)</td>\n",
       "      <td>195,052 (233)</td>\n",
       "      <td>78,524 (145)</td>\n",
       "      <td>0.7957 (0.0002)</td>\n",
       "      <td>0.6105 (0.0005)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG004</td>\n",
       "      <td>237,951 (490)</td>\n",
       "      <td>270,323 (490)</td>\n",
       "      <td>60,461 (365)</td>\n",
       "      <td>0.7974 (0.0007)</td>\n",
       "      <td>0.4682 (0.0010)</td>\n",
       "      <td>273,779 (416)</td>\n",
       "      <td>234,495 (416)</td>\n",
       "      <td>71,535 (257)</td>\n",
       "      <td>0.7929 (0.0004)</td>\n",
       "      <td>0.5386 (0.0008)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample  Imputed 4P2X TP  Imputed 4P2X FN Imputed 4P2X FP  \\\n",
       "0  HG002  251,862 (1,203)  270,527 (1,203)    65,080 (614)   \n",
       "1  HG003    256,262 (716)    244,530 (716)    70,962 (416)   \n",
       "2  HG004    237,951 (490)    270,323 (490)    60,461 (365)   \n",
       "\n",
       "  Imputed 4P2X Precision Imputed 4P2X Recall Imputed 8P5X TP Imputed 8P5X FN  \\\n",
       "0        0.7947 (0.0008)     0.4821 (0.0023)   294,850 (254)   227,539 (254)   \n",
       "1        0.7832 (0.0005)     0.5117 (0.0014)   305,742 (233)   195,052 (233)   \n",
       "2        0.7974 (0.0007)     0.4682 (0.0010)   273,779 (416)   234,495 (416)   \n",
       "\n",
       "  Imputed 8P5X FP Imputed 8P5X Precision Imputed 8P5X Recall  \n",
       "0    76,268 (112)        0.7945 (0.0002)     0.5644 (0.0005)  \n",
       "1    78,524 (145)        0.7957 (0.0002)     0.6105 (0.0005)  \n",
       "2    71,535 (257)        0.7929 (0.0004)     0.5386 (0.0008)  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "    'Sample': [],\n",
    "    'Imputed 4P2X TP': [],\n",
    "    'Imputed 4P2X FN': [],\n",
    "    'Imputed 4P2X FP': [],\n",
    "    'Imputed 4P2X Precision': [],\n",
    "    'Imputed 4P2X Recall': [],\n",
    "    'Imputed 8P5X TP': [],\n",
    "    'Imputed 8P5X FN': [],\n",
    "    'Imputed 8P5X FP': [],\n",
    "    'Imputed 8P5X Precision': [],\n",
    "    'Imputed 8P5X Recall': [],\n",
    "}\n",
    "\n",
    "for hgid, naid in hgid2naid.items():\n",
    "    data['Sample'].append(hgid)\n",
    "    \n",
    "    df_temp = df_glimpse_variants[df_glimpse_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 4) & (df_temp.WGS_depth == 2)]\n",
    "    data['Imputed 4P2X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['Imputed 4P2X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['Imputed 4P2X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['Imputed 4P2X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['Imputed 4P2X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    # print(hgid, 'Imputed WEGS 4P2X', len(df_temp))\n",
    "    \n",
    "    df_temp = df_glimpse_variants[df_glimpse_variants.HGID == hgid]\n",
    "    df_temp = df_temp[(df_temp.Type == 'INDEL') & (df_temp.Filter == 'ALL') & (df_temp.Plexing == 8) & (df_temp.WGS_depth == 5)]\n",
    "    data['Imputed 8P5X TP'].append(f'{np.mean(df_temp.TP):,.0f} ({stats.sem(df_temp.TP):,.0f})')\n",
    "    data['Imputed 8P5X FP'].append(f'{np.mean(df_temp.FP):,.0f} ({stats.sem(df_temp.FP):,.0f})')\n",
    "    data['Imputed 8P5X FN'].append(f'{np.mean(df_temp.FN):,.0f} ({stats.sem(df_temp.FN):,.0f})')\n",
    "    data['Imputed 8P5X Precision'].append(f'{np.mean(df_temp.Precision):.4f} ({stats.sem(df_temp.Precision):,.4f})')\n",
    "    data['Imputed 8P5X Recall'].append(f'{np.mean(df_temp.Recall):.4f} ({stats.sem(df_temp.Recall):,.4f})')\n",
    "    # print(hgid, 'Imputed WEGS 8P5X', len(df_temp))\n",
    "    \n",
    "pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d02cf88-e901-4c04-9293-61fc91b45f0e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
